
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="dmeth: A toolkit for transparent, reproducible DNA methylation analysis">
      
      
        <meta name="author" content="Dare Afolabi">
      
      
        <link rel="canonical" href="https://dare-afolabi.github.io/dmeth/UserGuide/">
      
      
        <link rel="prev" href="..">
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>User Guide - dmeth Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.618322db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#dmeth-documentation-for-reproducible-dna-methylation-analysis" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="dmeth Documentation" class="md-header__button md-logo" aria-label="dmeth Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            dmeth Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              User Guide
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/dare-afolabi/dmeth" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="dmeth Documentation" class="md-nav__button md-logo" aria-label="dmeth Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    dmeth Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/dare-afolabi/dmeth" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    User Guide
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    User Guide
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#user-guide" class="md-nav__link">
    <span class="md-ellipsis">
      
        User Guide
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#key-features" class="md-nav__link">
    <span class="md-ellipsis">
      
        Key Features
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#quick-start" class="md-nav__link">
    <span class="md-ellipsis">
      
        Quick Start
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Installation
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reference-for-all-modules-classes-and-functions" class="md-nav__link">
    <span class="md-ellipsis">
      
        Reference for all modules, classes, and functions
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preprocessing-utilities" class="md-nav__link">
    <span class="md-ellipsis">
      
        Preprocessing Utilities
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preprocessing Utilities">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#attributes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Attributes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_6" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_7" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_6" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#study-planning" class="md-nav__link">
    <span class="md-ellipsis">
      
        Study Planning
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Study Planning">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#features_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_8" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_9" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_7" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_10" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_8" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_11" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_9" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_10" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_12" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_11" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_13" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_12" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_14" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_13" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_15" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_14" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_15" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_16" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_16" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_17" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_18" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#core-differential-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      
        Core Differential Analysis
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Core Differential Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#core-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      
        Core Analysis
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_19" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_17" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_20" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_18" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_21" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_19" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#postprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      
        Postprocessing
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_22" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_20" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_6" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_23" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_21" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_6" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#preparation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Preparation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_24" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_22" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_7" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_25" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_23" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_8" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_26" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_24" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_7" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#validation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Validation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_27" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_25" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_9" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_28" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_26" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_10" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_29" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_27" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_11" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_30" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_28" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_12" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#downstream-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      
        Downstream Analysis
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Downstream Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Annotation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_6" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_31" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_29" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_32" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_30" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_8" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_33" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_31" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_34" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_32" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_35" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_33" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_13" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_36" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_34" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#deconvolution" class="md-nav__link">
    <span class="md-ellipsis">
      
        Deconvolution
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_7" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_37" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_35" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_14" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_9" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#downstream-stats" class="md-nav__link">
    <span class="md-ellipsis">
      
        Downstream Stats
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_8" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_38" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_36" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_39" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_37" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_40" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_38" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_41" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_39" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_42" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_40" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_43" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_41" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_44" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_42" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_15" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_45" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_43" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#helpers" class="md-nav__link">
    <span class="md-ellipsis">
      
        Helpers
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_9" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_46" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_44" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_10" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#signature" class="md-nav__link">
    <span class="md-ellipsis">
      
        Signature
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_10" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_47" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_45" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_48" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_46" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_11" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_49" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_47" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_12" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#configuration" class="md-nav__link">
    <span class="md-ellipsis">
      
        Configuration
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Configuration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config-manager" class="md-nav__link">
    <span class="md-ellipsis">
      
        Config Manager
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_11" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_50" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_48" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_51" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_16" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_52" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_49" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_17" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_53" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_54" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_18" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#inputoutput" class="md-nav__link">
    <span class="md-ellipsis">
      
        Input/Output
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Input/Output">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#data-utils" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Utils
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#key-components" class="md-nav__link">
    <span class="md-ellipsis">
      
        Key Components
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Key Components">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#parameters_55" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#attributes_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Attributes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_13" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#readers" class="md-nav__link">
    <span class="md-ellipsis">
      
        Readers
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_12" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_56" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#attributes_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Attributes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_14" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_57" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_50" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_15" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_58" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_51" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_19" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#writers" class="md-nav__link">
    <span class="md-ellipsis">
      
        Writers
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_13" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_59" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#attributes_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Attributes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_16" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_60" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_52" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_61" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_53" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_20" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_62" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_21" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#supported-formats" class="md-nav__link">
    <span class="md-ellipsis">
      
        Supported formats
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Supported formats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#parameters_63" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_22" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utilities" class="md-nav__link">
    <span class="md-ellipsis">
      
        Utilities
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Utilities">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        Logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_14" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_64" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_54" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting" class="md-nav__link">
    <span class="md-ellipsis">
      
        Plotting
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_15" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_65" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#attributes_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Attributes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_17" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_66" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_55" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_67" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_56" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_68" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_18" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_69" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_57" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_70" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_58" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_71" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_59" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_72" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_60" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_73" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_61" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_74" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_62" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_75" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_63" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Citation
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Citation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      
        References
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#support" class="md-nav__link">
    <span class="md-ellipsis">
      
        Support
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#user-guide" class="md-nav__link">
    <span class="md-ellipsis">
      
        User Guide
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#key-features" class="md-nav__link">
    <span class="md-ellipsis">
      
        Key Features
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#quick-start" class="md-nav__link">
    <span class="md-ellipsis">
      
        Quick Start
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Installation
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reference-for-all-modules-classes-and-functions" class="md-nav__link">
    <span class="md-ellipsis">
      
        Reference for all modules, classes, and functions
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preprocessing-utilities" class="md-nav__link">
    <span class="md-ellipsis">
      
        Preprocessing Utilities
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preprocessing Utilities">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#attributes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Attributes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_6" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_7" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_6" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#study-planning" class="md-nav__link">
    <span class="md-ellipsis">
      
        Study Planning
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Study Planning">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#features_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_8" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_9" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_7" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_10" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_8" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_11" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_9" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_10" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_12" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_11" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_13" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_12" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_14" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_13" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_15" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_14" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_15" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_16" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_16" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_17" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_18" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#core-differential-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      
        Core Differential Analysis
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Core Differential Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#core-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      
        Core Analysis
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_19" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_17" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_20" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_18" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_21" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_19" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#postprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      
        Postprocessing
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_22" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_20" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_6" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_23" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_21" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_6" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#preparation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Preparation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_24" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_22" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_7" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_25" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_23" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_8" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_26" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_24" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_7" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#validation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Validation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_5" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_27" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_25" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_9" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_28" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_26" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_10" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_29" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_27" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_11" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_30" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_28" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_12" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#downstream-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      
        Downstream Analysis
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Downstream Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Annotation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_6" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_31" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_29" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_32" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_30" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_8" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_33" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_31" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_34" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_32" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_35" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_33" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_13" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_36" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_34" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#deconvolution" class="md-nav__link">
    <span class="md-ellipsis">
      
        Deconvolution
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_7" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_37" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_35" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_14" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_9" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#downstream-stats" class="md-nav__link">
    <span class="md-ellipsis">
      
        Downstream Stats
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_8" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_38" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_36" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_39" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_37" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_40" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_38" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_41" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_39" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_42" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_40" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_43" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_41" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_44" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_42" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_15" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_45" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_43" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#helpers" class="md-nav__link">
    <span class="md-ellipsis">
      
        Helpers
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_9" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_46" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_44" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_10" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#signature" class="md-nav__link">
    <span class="md-ellipsis">
      
        Signature
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_10" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_47" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_45" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_48" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_46" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_11" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_49" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_47" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_12" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#configuration" class="md-nav__link">
    <span class="md-ellipsis">
      
        Configuration
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Configuration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#config-manager" class="md-nav__link">
    <span class="md-ellipsis">
      
        Config Manager
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_11" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_50" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_48" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_51" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_16" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_52" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_49" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_17" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_53" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_54" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_18" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#inputoutput" class="md-nav__link">
    <span class="md-ellipsis">
      
        Input/Output
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Input/Output">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#data-utils" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Utils
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#key-components" class="md-nav__link">
    <span class="md-ellipsis">
      
        Key Components
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Key Components">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#parameters_55" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#attributes_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Attributes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_13" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#readers" class="md-nav__link">
    <span class="md-ellipsis">
      
        Readers
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_12" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_56" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#attributes_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        Attributes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_14" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_57" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_50" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_15" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_58" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_51" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_19" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#writers" class="md-nav__link">
    <span class="md-ellipsis">
      
        Writers
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_13" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_59" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#attributes_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        Attributes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_16" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_60" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_52" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_61" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_53" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_20" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_62" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_21" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#supported-formats" class="md-nav__link">
    <span class="md-ellipsis">
      
        Supported formats
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Supported formats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#parameters_63" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#raises_22" class="md-nav__link">
    <span class="md-ellipsis">
      
        Raises
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utilities" class="md-nav__link">
    <span class="md-ellipsis">
      
        Utilities
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Utilities">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#logger" class="md-nav__link">
    <span class="md-ellipsis">
      
        Logger
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_14" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_64" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_54" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting" class="md-nav__link">
    <span class="md-ellipsis">
      
        Plotting
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#features_15" class="md-nav__link">
    <span class="md-ellipsis">
      
        Features
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_65" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#attributes_4" class="md-nav__link">
    <span class="md-ellipsis">
      
        Attributes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_17" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_66" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_55" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_67" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_56" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_68" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#notes_18" class="md-nav__link">
    <span class="md-ellipsis">
      
        Notes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_69" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_57" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_70" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_58" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_71" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_59" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_72" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_60" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_73" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_61" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_74" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_62" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_75" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#returns_63" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Citation
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Citation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      
        References
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#support" class="md-nav__link">
    <span class="md-ellipsis">
      
        Support
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="dmeth-documentation-for-reproducible-dna-methylation-analysis">dmeth - documentation for reproducible DNA methylation analysis<a class="headerlink" href="#dmeth-documentation-for-reproducible-dna-methylation-analysis" title="Permanent link">&para;</a></h1>
<h2 id="user-guide">User Guide<a class="headerlink" href="#user-guide" title="Permanent link">&para;</a></h2>
<div align="center">
  <a href="https://codecov.io/gh/dare-afolabi/dmeth">
    <img src="https://img.shields.io/codecov/c/github/dare-afolabi/dmeth?style=flat" alt="Coverage">
  </a>
  <a href="https://www.python.org/downloads/">
    <img src="https://img.shields.io/badge/python-3.9+-blue.svg" alt="Python 3.9+">
  </a>
  <a href="https://badge.fury.io/py/dmeth">
    <img src="https://badge.fury.io/py/dmeth.svg" alt="PyPI version">
  </a>
  <a href="https://github.com/sponsors/dare-afolabi">
    <img src="https://img.shields.io/badge/Sponsor-grey?style=flat&logo=github-sponsors" alt="Sponsor">
  </a>
</div>

<p>A fast, statistically rigorous Python framework providing a toolkit for DNA methylation analysis - from raw beta matrices to biomarkers and functional interpretation. <strong><code>dmeth</code></strong> implements the full modern differential methylation pipeline used in high-impact epigenome-wide association studies (EWAS), with performance and correctness on par with established R/bioconductor tools, all in pure Python.</p>
<h2 id="key-features">Key Features<a class="headerlink" href="#key-features" title="Permanent link">&para;</a></h2>
<table>
<thead>
<tr>
<th>Feature</th>
<th>Implementation</th>
<th>Performance</th>
</tr>
</thead>
<tbody>
<tr>
<td>Empirical Bayes moderated t-tests</td>
<td>limma-style (Smyth 2004) with exact replication</td>
<td>Numba-accelerated (10100 faster)</td>
</tr>
<tr>
<td>Memory-efficient chunked analysis</td>
<td>Automatic fallback for &gt;1M probes</td>
<td>&lt;4 GB RAM typical</td>
</tr>
<tr>
<td>Cell-type deconvolution</td>
<td>Reference-based NNLS (Houseman/Horvath-style)</td>
<td>Parallel joblib</td>
</tr>
<tr>
<td>DMR discovery</td>
<td>Sliding-window clustering + gap merging</td>
<td>Vectorized</td>
</tr>
<tr>
<td>Gene annotation &amp; pathway enrichment</td>
<td>IntervalTree + Fishers exact (FDR)</td>
<td>Sub-second on 450k/EPIC</td>
</tr>
<tr>
<td>Coordinate liftover (hg19  hg38)</td>
<td>pyliftover integration</td>
<td>Per-region tracking</td>
</tr>
<tr>
<td>Biomarker panel discovery &amp; validation</td>
<td>RF / Elastic Net + stratified CV</td>
<td>Built-in</td>
</tr>
<tr>
<td>Robust preprocessing &amp; QC</td>
<td>Missingness, group representation, imputation</td>
<td>Production-safe</td>
</tr>
</tbody>
</table>
<p>Fully supports <strong>Illumina 450K</strong>, <strong>EPIC (850K)</strong>, and any custom CpG  sample matrix.</p>
<h2 id="quick-start">Quick Start<a class="headerlink" href="#quick-start" title="Permanent link">&para;</a></h2>
<div class="codehilite"><pre><span></span><code>pip<span class="w"> </span>install<span class="w"> </span><span class="s2">&quot;dmeth[full]&quot;</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">dmeth.io.readers</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_methylation_data</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">dmeth.core.analysis.preparation</span><span class="w"> </span><span class="kn">import</span> <span class="n">filter_cpgs_by_missingness</span><span class="p">,</span> <span class="n">impute_missing_values</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">dmeth.core.analysis.validation</span><span class="w"> </span><span class="kn">import</span> <span class="n">validate_design</span><span class="p">,</span> <span class="n">validate_contrast</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">dmeth.core.analysis.core_analysis</span><span class="w"> </span><span class="kn">import</span> <span class="n">fit_differential</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">dmeth.core.downstream.annotation</span><span class="w"> </span><span class="kn">import</span> <span class="n">find_dmrs_by_sliding_window</span>

<span class="c1"># 1. Load your data</span>
<span class="c1"># beta: CpG x samples matrix</span>
<span class="c1"># pheno: sample metadata with a &#39;group&#39; column</span>
<span class="n">beta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;beta_matrix.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">pheno</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;phenotype.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># 2. Preprocessing</span>
<span class="c1"># Drop CpGs with too much missingness</span>
<span class="n">beta_clean</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">filter_cpgs_by_missingness</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span> <span class="n">max_missing_rate</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>

<span class="c1"># Impute remaining missing values (kNN)</span>
<span class="n">beta_imp</span> <span class="o">=</span> <span class="n">impute_missing_values</span><span class="p">(</span><span class="n">beta_clean</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;knn&quot;</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

<span class="c1"># 3. Differential analysis (case vs control)</span>
<span class="c1"># Build design matrix from phenotype</span>
<span class="n">design</span> <span class="o">=</span> <span class="n">validate_design</span><span class="p">(</span><span class="n">pheno</span><span class="p">[</span><span class="s2">&quot;group&quot;</span><span class="p">])</span>
<span class="n">contrast</span> <span class="o">=</span> <span class="n">validate_contrast</span><span class="p">(</span><span class="n">design</span><span class="p">,</span> <span class="s2">&quot;case-control&quot;</span><span class="p">)</span>

<span class="c1"># Fit</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">fit_differential</span><span class="p">(</span>
    <span class="n">M</span><span class="o">=</span><span class="n">beta_imp</span><span class="p">,</span>
    <span class="n">design</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">design</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">beta_imp</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span>
    <span class="n">contrast</span><span class="o">=</span><span class="n">contrast</span><span class="p">,</span>
    <span class="n">shrink</span><span class="o">=</span><span class="s2">&quot;smyth&quot;</span><span class="p">,</span>
    <span class="n">robust</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># 4. Discover DMRs</span>
<span class="n">annotation</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;cpg_annotation.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># must include chr, pos columns</span>
<span class="n">dmrs</span> <span class="o">=</span> <span class="n">find_dmrs_by_sliding_window</span><span class="p">(</span>
    <span class="n">dms</span><span class="o">=</span><span class="n">res</span><span class="p">[</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;padj&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">],</span>
    <span class="n">annotation</span><span class="o">=</span><span class="n">annotation</span><span class="p">,</span>
    <span class="n">max_gap</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
    <span class="n">min_cpgs</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">dmrs</span><span class="p">)</span><span class="si">}</span><span class="s2"> DMRs&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dmrs</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</code></pre></div>

<h2 id="installation">Installation<a class="headerlink" href="#installation" title="Permanent link">&para;</a></h2>
<div class="codehilite"><pre><span></span><code><span class="c1"># Minimal (no speed, annotation, and other extras)</span>
pip<span class="w"> </span>install<span class="w"> </span>dmeth

<span class="c1"># Recommended: full scientific environment</span>
pip<span class="w"> </span>install<span class="w"> </span><span class="s2">&quot;dmeth[full]&quot;</span>

<span class="c1"># Development</span>
pip<span class="w"> </span>install<span class="w"> </span><span class="s2">&quot;dmeth[full,dev]&quot;</span>
</code></pre></div>

<p>Optional extras (dmeth[full]):</p>
<ul>
<li><strong>speed</strong>: numba, combat (highly recommended)</li>
<li><strong>annotation</strong>: intervaltree, pyliftover</li>
<li><strong>parallel</strong>: joblib</li>
<li><strong>format</strong>: PyYAML, toml, h5py, xlrd</li>
<li><strong>plotting</strong>: plotly, umap-learn</li>
<li><strong>io</strong>: pyarrow, tables, openpyxl, xlsxwriter</li>
</ul>
<p>Optional dev extras (dmeth[dev]):</p>
<p>pytest, pytest-cov, black, isort, flake8, flake8-pyproject, flake8-bugbear, bandit, mypy, mkdocs, mkdocs-material</p>
<h2 id="reference-for-all-modules-classes-and-functions">Reference for all modules, classes, and functions<a class="headerlink" href="#reference-for-all-modules-classes-and-functions" title="Permanent link">&para;</a></h2>
<hr />
<h2 id="preprocessing-utilities">Preprocessing Utilities<a class="headerlink" href="#preprocessing-utilities" title="Permanent link">&para;</a></h2>
<p>Core preprocessing utilities for DNA methylation array data.</p>
<ul>
<li>This module implements quality control, normalization, batch correction,
and filtering operations for methylation matrices (beta or M-values).</li>
<li>It provides both standard and high-performance implementations with
memory-efficient processing for large datasets.</li>
</ul>
<h3 id="features">Features<a class="headerlink" href="#features" title="Permanent link">&para;</a></h3>
<ul>
<li>Sample-level QC: Flag/remove samples with excessive missing data</li>
<li>CpG-level QC: Remove probes with high missingness or on sex chromosomes</li>
<li>Normalization: Beta quantile normalization with optional M-value conversion</li>
<li>Batch correction: Regression-based or ComBat-style correction</li>
<li>Filtering: Remove low-variance probes</li>
</ul>
<hr />
<p><code>ProcessedData(M: DataFrame, pheno: DataFrame, ann: DataFrame, meta: Dict[str, Any] = &lt;factory&gt;) -&gt; None</code></p>
<p>Central container for aligned methylation data used throughout the DMeth pipeline.</p>
<p>Guarantees that:</p>
<ul>
<li>All components (methylation matrix, sample metadata, probe annotation)     share consistent string indices.</li>
<li>Sample and probe alignment is validated on construction.</li>
<li>A <code>meta</code> dictionary tracks processing history (normalisation,     batch correction, QC metrics, etc.).</li>
</ul>
<p>All downstream functions expect or return instances of this class.</p>
<h3 id="parameters">Parameters<a class="headerlink" href="#parameters" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix with CpG sites as rows and samples as columns.
    Values are typically beta (01) or M-values (-inf to +inf).</li>
<li><strong>pheno</strong>: <code>pd.DataFrame</code>
    Sample metadata table. Must be indexed by sample IDs (strings         after construction).</li>
<li><strong>ann</strong>: <code>pd.DataFrame or None</code>
    Optional probe/CpG annotation table (e.g., Illumina manifest).         Must be indexed by CpG IDs.
    If <code>None</code>, probe-level alignment checks are skipped.</li>
<li><strong>meta</strong>: <code>dict, optional</code>
    Free-form dictionary storing pipeline provenance and parameters.
    Pre-populated with sensible defaults if not provided.</li>
</ul>
<h3 id="attributes">Attributes<a class="headerlink" href="#attributes" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix (string-indexed rows and columns).</li>
<li><strong>pheno</strong>: <code>pd.DataFrame</code>
    Sample metadata (string-indexed).</li>
<li><strong>ann</strong>: <code>pd.DataFrame or None</code>
    Probe annotation (string-indexed if present).</li>
<li><strong>meta</strong>: <code>dict</code>
    Processing metadata (e.g., <code>{"matrix_type": "beta",         "normalized": True, ...}</code>).</li>
</ul>
<h3 id="notes">Notes<a class="headerlink" href="#notes" title="Permanent link">&para;</a></h3>
<p>The <code>__post_init__</code> method automatically:</p>
<ul>
<li>Converts all relevant indices to strings.</li>
<li>Validates alignment between components.</li>
<li>Raises informative <code>KeyError</code> on mismatch.</li>
</ul>
<hr />
<p><code>batch_correction(data: ProcessedData, batch_col: str, covariates: Optional[List[str]] = None, method: str = 'qr', weights: Optional[ndarray] = None, block_size: int = 50000, robust: bool = False, return_diagnostics: bool = False) -&gt; Union[ProcessedData, Tuple[ProcessedData, Dict[str, Any]]]</code></p>
<p>Regression-based batch correction using full linear model (intercept +     covariates + batch).</p>
<p>Numerically stable QR/pinv solver with optional per-probe weights and     block-wise processing.</p>
<h3 id="parameters_1">Parameters<a class="headerlink" href="#parameters_1" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>batch_col</strong>: <code>str</code>
    Column in <code>pheno</code> identifying batch.</li>
<li><strong>covariates</strong>: <code>list[str] or None</code>
    Additional technical or biological covariates to preserve.</li>
<li><strong>method</strong>: <code>{"qr", "pinv"}, default "qr"</code>
    Solver backend.</li>
<li><strong>weights</strong>: <code>ndarray or None</code>
    Per-probe inverse-variance weights.</li>
<li><strong>robust</strong>: <code>bool, default False</code>
    Use Huber robust regression (much slower).</li>
<li><strong>return_diagnostics</strong>: <code>bool, default False</code>
    Return detailed effect sizes, residuals, and variance explained.</li>
</ul>
<h3 id="returns">Returns<a class="headerlink" href="#returns" title="Permanent link">&para;</a></h3>
<p>ProcessedData or (ProcessedData, dict)
    Corrected data and optional diagnostics.</p>
<hr />
<p><code>batch_correction_combat(data: ProcessedData, batch_col: str, covariates: Optional[List[str]] = None, parametric: bool = True, return_diagnostics: bool = False) -&gt; Union[ProcessedData, Tuple[ProcessedData, Dict[str, Any]]]</code></p>
<p>Wrapper for ComBat (or safe fallback) batch correction.</p>
<ul>
<li>Uses <code>pycombat</code> if installed (parametric or nonparametric).</li>
<li>Falls back to conservative mean-centering + variance rescaling with clear warning.</li>
<li>Automatically handles beta  M-value conversion.</li>
</ul>
<h3 id="parameters_2">Parameters<a class="headerlink" href="#parameters_2" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>parametric</strong>: <code>bool, default True</code>
    Use parametric ComBat (faster, assumes normality).</li>
<li><strong>return_diagnostics</strong>: <code>bool, default False</code>
    Return metadata about which method was actually used.</li>
</ul>
<h3 id="returns_1">Returns<a class="headerlink" href="#returns_1" title="Permanent link">&para;</a></h3>
<p>ProcessedData or (ProcessedData, dict)
    Batch-corrected data and diagnostics.</p>
<hr />
<p><code>filter_low_variance_cpgs(data: ProcessedData, min_percentile: float = 10.0, inplace: bool = True) -&gt; ProcessedData</code></p>
<p>Remove CpGs below a variance percentile threshold (e.g., bottom 10%).</p>
<h3 id="parameters_3">Parameters<a class="headerlink" href="#parameters_3" title="Permanent link">&para;</a></h3>
<ul>
<li>data : ProcessedData</li>
<li>min_percentile : float, default 10.0
    Keep only probes with variance  this percentile.</li>
<li>inplace : bool, default True
    Modify input object or return a copy.</li>
</ul>
<h3 id="returns_2">Returns<a class="headerlink" href="#returns_2" title="Permanent link">&para;</a></h3>
<ul>
<li>ProcessedData
    Dataset with low-variance probes removed.</li>
</ul>
<h3 id="notes_1">Notes<a class="headerlink" href="#notes_1" title="Permanent link">&para;</a></h3>
<ul>
<li>Number of removed probes recorded in <code>data.meta["qc"]["low_variance_removed"]</code>.</li>
</ul>
<hr />
<p><code>normalize_methylation(data: ProcessedData, method: str = 'beta_quantile', convert_to: Optional[str] = None, copy: bool = False, q_chunk_threshold: int = 100000000.0, q_block_probes: int = 50000) -&gt; ProcessedData</code></p>
<p>Perform quantile normalization across samples with optional beta      M-value conversion.</p>
<ul>
<li>Implements memory-efficient beta quantile normalization using column-wise     sorting and block-wise rank mapping.</li>
<li>Automatically falls back to disk-backed memmap for very large datasets.</li>
</ul>
<h3 id="parameters_4">Parameters<a class="headerlink" href="#parameters_4" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>data</strong>: <code>ProcessedData</code>
    Input methylation data.</li>
<li><strong>method</strong>: <code>{"beta_quantile", "none"}, default "beta_quantile"</code>
    Normalization method.</li>
<li><strong>convert_to</strong>: <code>{"beta", "m"} or None, optional</code>
    Convert matrix type after normalization (e.g., "m" for M-values).</li>
<li><strong>copy</strong>: <code>bool, default False</code>
    Work on a deep copy of the input.</li>
<li><strong>q_chunk_threshold</strong>: <code>int, default 1e8</code>
    Element count threshold (n_probes  n_samples) above which memmap is used.</li>
<li><strong>q_block_probes</strong>: <code>int, default 50_000</code>
    Number of probes processed per block during rank-to-target assignment.</li>
</ul>
<h3 id="returns_3">Returns<a class="headerlink" href="#returns_3" title="Permanent link">&para;</a></h3>
<p>ProcessedData
    Normalized dataset with updated <code>meta["normalized"]</code> and detailed provenance.</p>
<h3 id="notes_2">Notes<a class="headerlink" href="#notes_2" title="Permanent link">&para;</a></h3>
<p>Preserves original NaNs. Records full normalization metadata including     memory strategy.</p>
<hr />
<p><code>normalize_methylation_highperf(data: ProcessedData, method: str = 'beta_quantile', convert_to: Optional[str] = None, copy: bool = False, memmap_threshold: int = 200000000.0, memmap_dir: Optional[str] = None, n_workers: Optional[int] = None, sample_block: int = 16, random_state: Optional[int] = None) -&gt; ProcessedData</code></p>
<p>High-performance quantile normalization with multiprocessing and automatic     memmap handling.</p>
<p>Designed for &gt;850k  500+ datasets. Uses disk-backed arrays when needed     and parallelizes the rank-mapping stage.</p>
<h3 id="parameters_5">Parameters<a class="headerlink" href="#parameters_5" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>memmap_threshold</strong>: <code>int, default 2e8</code>
    Element count above which temporary memmap files are created.</li>
<li><strong>memmap_dir</strong>: <code>str or None</code>
    Directory for memmap files (defaults to system temp).</li>
<li><strong>n_workers</strong>: <code>int or None</code>
    Number of processes for parallel assignment (defaults to CPU count  1).</li>
<li><strong>sample_block</strong>: <code>int, default 16</code>
    Number of samples per parallel job.</li>
<li><strong>random_state</strong>: <code>int or None</code>
    Currently unused (kept for API consistency).</li>
</ul>
<h3 id="returns_4">Returns<a class="headerlink" href="#returns_4" title="Permanent link">&para;</a></h3>
<p>ProcessedData
    Quantile-normalized data with comprehensive metadata.</p>
<h3 id="notes_3">Notes<a class="headerlink" href="#notes_3" title="Permanent link">&para;</a></h3>
<p>Significantly faster and lower peak RAM than <code>normalize_methylation</code>     on large cohorts.</p>
<hr />
<p><code>qc_cpg_level(data: ProcessedData, max_missing_fraction: float = 0.1, drop_sex_chr: bool = True, chr_col: str = 'chromosome') -&gt; ProcessedData</code></p>
<p>Remove CpGs with high missingness across samples and optionally drop     sex-chromosome probes.</p>
<h3 id="parameters_6">Parameters<a class="headerlink" href="#parameters_6" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>data</strong>: <code>ProcessedData</code>
    Input dataset.</li>
<li><strong>max_missing_fraction</strong>: <code>float, default 0.10</code>
    Maximum fraction of samples allowed to be missing for a CpG (10% default).</li>
<li><strong>drop_sex_chr</strong>: <code>bool, default True</code>
    If True and annotation is available, remove all probes on chromosomes X and Y.</li>
<li><strong>chr_col</strong>: <code>str, default "chromosome"</code>
    Column name in <code>data.ann</code> containing chromosome information.</li>
</ul>
<h3 id="returns_5">Returns<a class="headerlink" href="#returns_5" title="Permanent link">&para;</a></h3>
<p>ProcessedData
    Updated container with low-quality and/or sex-chromosome CpGs removed.</p>
<h3 id="notes_4">Notes<a class="headerlink" href="#notes_4" title="Permanent link">&para;</a></h3>
<p>Summary statistics are recorded in <code>data.meta["qc"]["cpg_missing"]</code>.</p>
<hr />
<p><code>qc_sample_level(data: ProcessedData, max_missing_fraction: float = 0.05, min_nonmissing_probes: Optional[int] = None, remove_samples: bool = True) -&gt; ProcessedData</code></p>
<p>Identify and optionally remove samples with excessive missing methylation data.</p>
<h3 id="parameters_7">Parameters<a class="headerlink" href="#parameters_7" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>data</strong>: <code>ProcessedData</code>
    Input methylation dataset.</li>
<li><strong>max_missing_fraction</strong>: <code>float, default 0.05</code>
    Maximum allowed fraction of missing CpGs per sample (5% by default).</li>
<li><strong>min_nonmissing_probes</strong>: <code>int or None, optional</code>
    Alternative absolute threshold: minimum number of detected (non-missing)         probes required.</li>
<li><strong>remove_samples</strong>: <code>bool, default True</code>
    If True, flagged samples are removed from both <code>M</code> and <code>pheno</code>;         if False, only metadata is recorded.</li>
</ul>
<h3 id="returns_6">Returns<a class="headerlink" href="#returns_6" title="Permanent link">&para;</a></h3>
<p>ProcessedData
    Updated container (samples removed if requested).</p>
<h3 id="notes_5">Notes<a class="headerlink" href="#notes_5" title="Permanent link">&para;</a></h3>
<p>Flagged sample list and thresholds are stored in <code>data.meta["qc"]    ["sample_missing"]</code>.</p>
<hr />
<h2 id="study-planning">Study Planning<a class="headerlink" href="#study-planning" title="Permanent link">&para;</a></h2>
<p>High-level planning API for DNA methylation studies.</p>
<p>This module provides a clean, user-friendly interface to the full study planning
capabilities of <code>DMeth</code>. It exposes all configuration queries, cost/timeline
estimators, and sample size calculators as simple top-level functions.</p>
<ul>
<li>All functions automatically use the global <code>PlannerConfig</code> singleton</li>
<li>(managed in <code>dmeth.config.config_manager</code>), so configuration loaded once is immediately available everywhere.</li>
</ul>
<h3 id="features_1">Features<a class="headerlink" href="#features_1" title="Permanent link">&para;</a></h3>
<ul>
<li>Direct function access: <code>list_platforms()</code>, <code>calculate_sample_size()</code>, etc.</li>
<li>Zero-boilerplate: no need to manually instantiate or pass config objects</li>
<li>Real-time updates: any config change (e.g., regional pricing) instantly affects all calculations</li>
</ul>
<hr />
<p><code>add_custom_platform(platform_id: str, platform_info: Dict[str, Any]) -&gt; None</code></p>
<p>Register a new custom platform after schema validation.</p>
<h3 id="parameters_8">Parameters<a class="headerlink" href="#parameters_8" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>platform_id</strong>: <code>str</code>
    Unique identifier for the new platform.</li>
<li><strong>platform_info</strong>: <code>dict</code>
    Platform data conforming to <code>PlatformSchema</code>.</li>
</ul>
<h3 id="raises">Raises<a class="headerlink" href="#raises" title="Permanent link">&para;</a></h3>
<p>ValueError
    If the provided data fails validation.</p>
<hr />
<p><code>calculate_sample_size(*args, **kwargs)</code></p>
<p>Compute required sample size per group using power analysis.</p>
<ul>
<li>The calculation accounts for the chosen experimental design, platform</li>
<li>CpG count (for Bonferroni correction), and any design-specific power
adjustment factor.</li>
</ul>
<h3 id="parameters_9">Parameters<a class="headerlink" href="#parameters_9" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>design_id</strong>: <code>str</code>
    Identifier of the experimental design.</li>
<li><strong>platform_id</strong>: <code>str</code>
    Identifier of the methylation platform.</li>
<li><strong>effect_size</strong>: <code>float</code>
    Expected Cohen's d.</li>
<li><strong>alpha</strong>: <code>float, default 0.05</code>
    Nominal type-I error rate.</li>
<li><strong>power</strong>: <code>float, default 0.8</code>
    Desired statistical power.</li>
<li><strong>mcp_method</strong>: <code>str, optional</code>
    Multiple-comparison correction method (<code>bonferroni</code>, <code>fdr</code>, <code>none</code>).</li>
<li><strong>max_n</strong>: <code>int, default 500</code>
    Upper bound on calculated sample size per group.</li>
</ul>
<h3 id="returns_7">Returns<a class="headerlink" href="#returns_7" title="Permanent link">&para;</a></h3>
<p>dict
    Contains <code>n_per_group</code>, <code>total_samples</code>, effective alpha, etc.</p>
<hr />
<p><code>estimate_study_timeline(*args, **kwargs)</code></p>
<p>Produce a detailed timeline estimate for the entire study.</p>
<p>The estimate respects phase-specific scaling, batch adjustments,
parallelisability, and an optional global contingency buffer.</p>
<h3 id="parameters_10">Parameters<a class="headerlink" href="#parameters_10" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>n_samples</strong>: <code>int</code>
    Total number of samples to be processed.</li>
<li><strong>platform_id</strong>: <code>str</code>
    Platform identifier (affects array-processing duration).</li>
<li><strong>contingency</strong>: <code>bool, default True</code>
    Apply the global contingency buffer percentage.</li>
</ul>
<h3 id="returns_8">Returns<a class="headerlink" href="#returns_8" title="Permanent link">&para;</a></h3>
<p>DataFrame
    One row per phase plus a summary <code>TOTAL</code> row.</p>
<hr />
<p><code>estimate_total_cost(*args, **kwargs)</code></p>
<p>Calculate the complete study cost and provide a detailed breakdown.</p>
<p>Handles per-sample, per-CpG, and fixed cost components, with optional
inclusion of components marked as <code>optional</code>.</p>
<h3 id="parameters_11">Parameters<a class="headerlink" href="#parameters_11" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>n_samples</strong>: <code>int</code>
    Number of samples.</li>
<li><strong>platform_id</strong>: <code>str</code>
    Platform identifier.</li>
<li><strong>include_optional</strong>: <code>bool, default True</code>
    Include optional cost components.</li>
<li><strong>validate</strong>: <code>bool, default True</code>
    Perform unit validation (kept for backward compatibility).</li>
</ul>
<h3 id="returns_9">Returns<a class="headerlink" href="#returns_9" title="Permanent link">&para;</a></h3>
<p>dict
    Contains a <code>components</code> DataFrame, <code>total</code> cost,
    <code>per_sample</code> cost, and metadata.</p>
<hr />
<p><code>get_config() -&gt; dmeth.config.config_manager.PlannerConfig</code></p>
<p>Retrieve the global singleton instance of <code>PlannerConfig</code>.</p>
<p>The instance is created on first access if necessary.</p>
<h3 id="returns_10">Returns<a class="headerlink" href="#returns_10" title="Permanent link">&para;</a></h3>
<p>PlannerConfig
    The active configuration object.</p>
<hr />
<p><code>get_cost_components(platform: Optional[str] = None, include_optional: bool = True) -&gt; Dict[str, Dict]</code></p>
<p>Return cost components, optionally filtered by platform or optionality.</p>
<h3 id="parameters_12">Parameters<a class="headerlink" href="#parameters_12" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>platform</strong>: <code>str, optional</code>
    Restrict to components whose <code>applies_to</code> list contains this platform.</li>
<li><strong>include_optional</strong>: <code>bool, default True</code>
    Whether to include components marked as optional.</li>
</ul>
<h3 id="returns_11">Returns<a class="headerlink" href="#returns_11" title="Permanent link">&para;</a></h3>
<p>dict
    Mapping from component identifier to component dictionary.</p>
<hr />
<p><code>get_design(design_id: str) -&gt; Dict[str, Any]</code></p>
<p>Retrieve the configuration dictionary for a specific experimental design.</p>
<h3 id="parameters_13">Parameters<a class="headerlink" href="#parameters_13" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>design_id</strong>: <code>str</code>
    Identifier of the design.</li>
</ul>
<h3 id="returns_12">Returns<a class="headerlink" href="#returns_12" title="Permanent link">&para;</a></h3>
<p>dict
    Design configuration.</p>
<h3 id="raises_1">Raises<a class="headerlink" href="#raises_1" title="Permanent link">&para;</a></h3>
<p>KeyError
    If the design identifier is unknown.</p>
<hr />
<p><code>get_platform(platform_id: str) -&gt; Dict[str, Any]</code></p>
<p>Retrieve the configuration dictionary for a specific platform.</p>
<h3 id="parameters_14">Parameters<a class="headerlink" href="#parameters_14" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>platform_id</strong>: <code>str</code>
    Identifier of the platform.</li>
</ul>
<h3 id="returns_13">Returns<a class="headerlink" href="#returns_13" title="Permanent link">&para;</a></h3>
<p>dict
    Platform configuration.</p>
<h3 id="raises_2">Raises<a class="headerlink" href="#raises_2" title="Permanent link">&para;</a></h3>
<p>KeyError
    If the platform identifier is unknown.</p>
<hr />
<p><code>get_platform_by_budget(max_cost_per_sample: float) -&gt; DataFrame</code></p>
<p>Return platforms whose per-sample cost does not exceed a budget limit.</p>
<h3 id="parameters_15">Parameters<a class="headerlink" href="#parameters_15" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>max_cost_per_sample</strong>: <code>float</code>
    Maximum acceptable cost per sample.</li>
</ul>
<h3 id="returns_14">Returns<a class="headerlink" href="#returns_14" title="Permanent link">&para;</a></h3>
<p>DataFrame
    Subset of platforms meeting the budget constraint.</p>
<hr />
<p><code>list_designs() -&gt; DataFrame</code></p>
<p>Return a table of all configured experimental designs.</p>
<h3 id="returns_15">Returns<a class="headerlink" href="#returns_15" title="Permanent link">&para;</a></h3>
<p>DataFrame
    One row per design.</p>
<hr />
<p><code>list_platforms(recommended_only: bool = False) -&gt; DataFrame</code></p>
<p>Return a DataFrame containing all configured platforms.</p>
<h3 id="parameters_16">Parameters<a class="headerlink" href="#parameters_16" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>recommended_only</strong>: <code>bool, default False</code>
    If True, only platforms with <code>recommended=True</code> are returned.</li>
</ul>
<h3 id="returns_16">Returns<a class="headerlink" href="#returns_16" title="Permanent link">&para;</a></h3>
<p>DataFrame
    One row per platform, indexed by platform identifier.</p>
<hr />
<p><code>update_platform_cost(platform_id: str, new_cost: float) -&gt; None</code></p>
<p>Change the per-sample cost of an existing platform.</p>
<h3 id="parameters_17">Parameters<a class="headerlink" href="#parameters_17" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>platform_id</strong>: <code>str</code>
    Platform identifier.</li>
<li><strong>new_cost</strong>: <code>float</code>
    New cost per sample (must be non-negative).</li>
</ul>
<h3 id="raises_3">Raises<a class="headerlink" href="#raises_3" title="Permanent link">&para;</a></h3>
<p>KeyError
    If the platform does not exist.</p>
<hr />
<p><code>update_regional_pricing(multiplier: float, region: Optional[str] = None) -&gt; None</code></p>
<p>Apply a uniform multiplicative adjustment to all platform costs.</p>
<h3 id="parameters_18">Parameters<a class="headerlink" href="#parameters_18" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>multiplier</strong>: <code>float</code>
    Factor by which current costs are multiplied (must be &gt; 0).</li>
<li><strong>region</strong>: <code>str, optional</code>
    Descriptive label for logging purposes.</li>
</ul>
<h3 id="raises_4">Raises<a class="headerlink" href="#raises_4" title="Permanent link">&para;</a></h3>
<p>ValueError
    If <code>multiplier</code> is not positive.</p>
<hr />
<h2 id="core-differential-analysis">Core Differential Analysis<a class="headerlink" href="#core-differential-analysis" title="Permanent link">&para;</a></h2>
<h3 id="core-analysis">Core Analysis<a class="headerlink" href="#core-analysis" title="Permanent link">&para;</a></h3>
<p>Statistical engines for differential methylation analysis.</p>
<p>This module implements empirical Bayes moderated statistics following
Smyth (2004) limma methodology, with extensions for paired designs,
robust estimation, and Numba-accelerated computation. It provides both
full-matrix and chunked processing for memory-constrained environments.</p>
<h3 id="features_2">Features<a class="headerlink" href="#features_2" title="Permanent link">&para;</a></h3>
<ul>
<li>Empirical Bayes variance shrinkage (Smyth method)</li>
<li>Robust variance estimation via winsorization</li>
<li>Numba JIT compilation for 10-100 x speedup</li>
<li>Memory-efficient chunked processing</li>
<li>Paired and multi-group designs</li>
<li>F-tests for multi-coefficient contrasts</li>
<li>Automatic handling of missing data</li>
<li>Group mean computation for interpretability</li>
</ul>
<hr />
<p><code>SmythPrior(df_prior: 'float', var_prior: 'float') -&gt; None</code></p>
<p>SmythPrior(df_prior: 'float', var_prior: 'float')</p>
<hr />
<p><code>check_analysis_memory(M: 'pd.DataFrame', warn_threshold_gb: 'float' = 8.0)</code></p>
<p>Conservatively estimate RAM requirements for a full-matrix differential analysis.</p>
<ul>
<li>Calculates current data footprint, projects peak usage (approximately 4 times     input size), and compares against available system memory.</li>
<li>Raises MemoryError early if danger is high, forcing use of chunked mode.</li>
</ul>
<h3 id="parameters_19">Parameters<a class="headerlink" href="#parameters_19" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix (CpGs  samples).</li>
<li><strong>warn_threshold_gb</strong>: <code>float, default 8.0</code>
    Issue a warning if estimated peak exceeds this value.</li>
</ul>
<h3 id="returns_17">Returns<a class="headerlink" href="#returns_17" title="Permanent link">&para;</a></h3>
<p>dict
    Keys: <code>data_gb</code>, <code>peak_gb</code>, <code>available_gb</code>.</p>
<h3 id="raises_5">Raises<a class="headerlink" href="#raises_5" title="Permanent link">&para;</a></h3>
<p>MemoryError
    If projected peak consumption exceeds ~80% of available RAM. In such cases         <code>fit_differential_chunked</code> should be used instead.</p>
<hr />
<p><code>fit_differential(M: 'pd.DataFrame', design: 'pd.DataFrame', contrast: 'Optional[np.ndarray]' = None, contrast_matrix: 'Optional[np.ndarray]' = None, shrink: 'Union[str, float]' = 'auto', robust: 'bool' = True, eps: 'float' = 1e-08, return_residuals: 'bool' = False, min_count: 'int' = 3, max_d0: 'float' = 50.0, winsor_lower: 'float' = 0.05, winsor_upper: 'float' = 0.95, group_labels: 'Optional[pd.Series]' = None, use_numba: 'bool' = True) -&gt; 'Union[pd.DataFrame, Tuple[pd.DataFrame, pd.DataFrame]]'</code></p>
<p>Core limma-style differential methylation analysis with empirical Bayes moderation.</p>
<ul>
<li><strong>Features</strong>: <code>- Numba-accelerated feature-wise fitting (10100 speedup when no missing data)</code></li>
<li>Automatic handling of missing values</li>
<li>Robust winsorization of residuals</li>
<li>Multiple shrinkage strategies (Smyth, median, fixed d, none)</li>
<li>Single-contrast t-tests or multi-coefficient F-tests</li>
<li>Optional group mean beta/M-value columns for interpretability</li>
<li>Optional residuals return for diagnostics</li>
</ul>
<h3 id="parameters_20">Parameters<a class="headerlink" href="#parameters_20" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    CpG  sample matrix of M-values (or beta values if already on log-odds scale).</li>
<li><strong>design</strong>: <code>pd.DataFrame</code>
    Sample  covariate design matrix (output from <code>patsy</code> or manual construction).
contrast / contrast_matrix : np.ndarray or None
    Contrast vector (t-test) or matrix (F-test).</li>
<li><strong>shrink</strong>: <code>{"auto", "smyth", "median", "none"} or float</code>
    Variance shrinkage method or fixed prior df.</li>
<li><strong>robust</strong>: <code>bool, default True</code>
    Winsorize residuals per feature.</li>
<li><strong>group_labels</strong>: <code>pd.Series, optional</code>
    Explicit sample grouping for mean columns; otherwise heuristically detected.</li>
<li><strong>use_numba</strong>: <code>bool, default True</code>
    Enable JIT-compiled loop when possible.</li>
</ul>
<h3 id="returns_18">Returns<a class="headerlink" href="#returns_18" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Results table with columns including <code>logFC</code>, <code>t</code>, <code>pval</code>, <code>padj</code>         (or <code>F</code>, <code>df1</code>, <code>df2</code>), moderated variances (<code>s2_post</code>), prior df         (<code>d0</code>), and group means.
(pd.DataFrame, pd.DataFrame), optional
    Second element contains per-feature residuals if <code>return_residuals=True</code>.</p>
<hr />
<p><code>fit_differential_chunked(M: 'pd.DataFrame', design: 'pd.DataFrame', chunk_size: 'int' = 10000, verbose: 'bool' = True, group_labels: 'Optional[pd.Series]' = None, **kwargs) -&gt; 'pd.DataFrame'</code></p>
<p>Memory-efficient version of <code>fit_differential</code> that processes features in chunks.</p>
<p>Useful for &gt;1M probe arrays on limited RAM. P-values are combined     globally and FDR-corrected once at the end.</p>
<h3 id="parameters_21">Parameters<a class="headerlink" href="#parameters_21" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>chunk_size</strong>: <code>int, default 10000</code>
    Number of CpGs per processing chunk.
**kwargs
    All additional arguments forwarded to <code>fit_differential</code>.</li>
</ul>
<h3 id="returns_19">Returns<a class="headerlink" href="#returns_19" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Complete results table with globally adjusted p-values and group means.</p>
<h3 id="postprocessing">Postprocessing<a class="headerlink" href="#postprocessing" title="Permanent link">&para;</a></h3>
<p>Postprocessing and interpretation utilities for differential methylation analysis results.</p>
<p>This module provides robust, publication-ready tools for summarizing, filtering, and extracting biologically meaningful differentially methylated CpGs from statistical output. It supports flexible significance criteria, effect-size thresholding, directional filtering, and delta-beta constraints, with comprehensive summary statistics for reporting.</p>
<h3 id="features_3">Features<a class="headerlink" href="#features_3" title="Permanent link">&para;</a></h3>
<ul>
<li>Comprehensive summary statistics including significance counts, directionality, effect sizes, and empirical Bayes shrinkage metrics</li>
<li>Flexible extraction of significant CpGs with adjustable logFC, adjusted p-value, and delta-beta thresholds</li>
<li>Directional filtering for hyper- and hypomethylated sites</li>
<li>Automatic detection or explicit specification of group mean beta-value columns for delta-beta filtering</li>
<li>Graceful handling of missing columns and empty result sets with informative warnings</li>
<li>Optional detailed summary dictionaries for downstream reporting or visualization</li>
</ul>
<hr />
<p><code>get_significant_cpgs(res: DataFrame, lfc_col: str = 'logFC', pval_col: str = 'padj', lfc_thresh: float = 0.0, pval_thresh: float = 0.05, delta_beta_thresh: Optional[float] = None, direction: Optional[str] = None, delta_beta_cols: Optional[List[str]] = None, return_summary: bool = False, verbose: bool = True) -&gt; Union[List[str], Dict]</code></p>
<p>Extract biologically meaningful significant CpGs using flexible,     multi-layer filtering.</p>
<ul>
<li>Combines adjusted p-value, log fold-change, directionality, and optional     absolute delta-beta criteria.</li>
<li>Ideal for generating final candidate lists for downstream validation or     pathway analysis.</li>
</ul>
<h3 id="parameters_22">Parameters<a class="headerlink" href="#parameters_22" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>res</strong>: <code>pd.DataFrame</code>
    Full differential methylation results table.</li>
<li><strong>lfc_col</strong>: <code>str, default "logFC"</code>
    Column containing log fold change.</li>
<li><strong>pval_col</strong>: <code>str, default "padj"</code>
    Column containing adjusted p-values.</li>
<li><strong>lfc_thresh</strong>: <code>float, default 0.0</code>
    Minimum absolute |logFC| required (0 means no LFC filtering).</li>
<li><strong>pval_thresh</strong>: <code>float, default 0.05</code>
    Maximum adjusted p-value for significance.</li>
<li><strong>delta_beta_thresh</strong>: <code>float or None, optional</code>
    Minimum absolute difference in mean beta values between groups.</li>
<li><strong>direction</strong>: <code>{"hyper", "hypo", None}, optional</code>
    Restrict to hypermethylated (logFC &gt; 0), hypomethylated (logFC &lt; 0), or both.</li>
<li><strong>delta_beta_cols</strong>: <code>list[str] or None, optional</code>
    Exactly two column names containing group mean beta values (e.g.         <code>["meanB_case", "meanB_control"]</code>).
    If <code>None</code> and delta-beta filtering is requested, automatically detects         columns starting with <code>meanB_</code>.</li>
<li><strong>return_summary</strong>: <code>bool, default False</code>
    If True, return a detailed dictionary instead of just the CpG list.</li>
<li><strong>verbose</strong>: <code>bool, default True</code>
    Warn when no significant CpGs are found.</li>
</ul>
<h3 id="returns_20">Returns<a class="headerlink" href="#returns_20" title="Permanent link">&para;</a></h3>
<p>list[str] or dict
    List of significant CpG IDs (default)
    Or a summary dictionary containing counts, directional breakdown,         mean |logFC|, and the CpG list</p>
<h3 id="raises_6">Raises<a class="headerlink" href="#raises_6" title="Permanent link">&para;</a></h3>
<p>KeyError
    If required columns (<code>lfc_col</code> or <code>pval_col</code>) are missing.
ValueError
    If <code>delta_beta_thresh</code> is used but exactly two mean-beta columns         cannot be identified.</p>
<hr />
<p><code>summarize_differential_results(res: DataFrame, pval_thresh: float = 0.05, verbose: bool = True) -&gt; Dict[str, Union[int, float]]</code></p>
<p>Produce a comprehensive, publication-ready summary of differential     methylation results.</p>
<p>Handles missing columns gracefully and returns key metrics for reporting:
significance counts, directionality, effect-size statistics, and     variance-shrinkage diagnostics.</p>
<h3 id="parameters_23">Parameters<a class="headerlink" href="#parameters_23" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>res</strong>: <code>pd.DataFrame</code>
    Differential methylation results containing at minimum <code>logFC</code> and <code>padj</code>.
    May also include <code>pval</code>, <code>s2</code>, <code>s2_post</code>, and <code>d0</code>         (prior degrees of freedom).</li>
<li><strong>pval_thresh</strong>: <code>float, default 0.05</code>
    Adjusted p-value threshold defining statistical significance.</li>
<li><strong>verbose</strong>: <code>bool, default True</code>
    Emit warnings for empty input or missing optional columns.</li>
</ul>
<h3 id="returns_21">Returns<a class="headerlink" href="#returns_21" title="Permanent link">&para;</a></h3>
<p>dict
    Summary dictionary with the following keys:
    <code>total_tested</code>: total CpGs tested
    <code>significant</code>: number of significant CpGs (padj &lt; threshold)
    <code>pct_significant</code>: percentage of significant CpGs
    <code>hypermethylated</code> / <code>hypomethylated</code>: directional counts
    <code>mean_abs_logFC_sig</code> / <code>median_abs_logFC_sig</code>:         effect size summaries among significant sites
    <code>max_abs_logFC</code>: largest absolute log fold change
    <code>min_pval</code>: smallest raw p-value
    <code>shrinkage_factor</code>: median ratio of moderated to original variance
    <code>d0</code>: median prior degrees of freedom (empirical Bayes)</p>
<h3 id="notes_6">Notes<a class="headerlink" href="#notes_6" title="Permanent link">&para;</a></h3>
<p>Missing optional columns are safely ignored with fallback values.
Empty input returns a zero-filled summary.</p>
<h3 id="preparation">Preparation<a class="headerlink" href="#preparation" title="Permanent link">&para;</a></h3>
<p>Data preparation and preprocessing utilities for differential methylation analysis.</p>
<p>This module provides robust, production-ready tools for cleaning and imputing
CpG-level methylation matrices prior to statistical modeling. It implements
stringent quality-control filters based on missingness patterns and group
representation, alongside flexible imputation strategies optimized for
high-dimensional epigenomic data.</p>
<h3 id="features_4">Features<a class="headerlink" href="#features_4" title="Permanent link">&para;</a></h3>
<ul>
<li>Global and per-group missingness filtering with configurable thresholds</li>
<li>Minimum representation enforcement across experimental groups</li>
<li>Multiple imputation methods: row-wise mean/median and K-nearest neighbors</li>
<li>Sample-wise KNN option for dramatic speed gains when samples  CpGs</li>
<li>Comprehensive input validation and clear diagnostic reporting</li>
<li>Memory-efficient operations using NumPy-based vectorization</li>
</ul>
<hr />
<p><code>filter_cpgs_by_missingness(M: 'pd.DataFrame', max_missing_rate: 'float' = 0.2, min_samples_per_group: 'Optional[int]' = None, groups: 'Optional[pd.Series]' = None) -&gt; 'Tuple[pd.DataFrame, int, int]'</code></p>
<p>Remove CpGs exceeding a global missingness threshold and/or failing     per-group representation.</p>
<h3 id="parameters_24">Parameters<a class="headerlink" href="#parameters_24" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix (CpGs  samples) containing possible NaN values.</li>
<li><strong>max_missing_rate</strong>: <code>float, default 0.2</code>
    Maximum allowable fraction of missing values across all samples (01).</li>
<li><strong>min_samples_per_group</strong>: <code>int or None, optional</code>
    Minimum number of non-missing observations required in every experimental group.
    Ignored if <code>None</code>.</li>
<li><strong>groups</strong>: <code>pd.Series or None, optional</code>
    Sample group labels indexed by <code>M.columns</code>. Required when         <code>min_samples_per_group</code> is set.</li>
</ul>
<h3 id="returns_22">Returns<a class="headerlink" href="#returns_22" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>filtered</strong>: <code>pd.DataFrame</code>
    Subset of CpGs passing both filters.</li>
<li><strong>n_removed</strong>: <code>int</code>
    Number of CpGs removed.</li>
<li><strong>n_kept</strong>: <code>int</code>
    Number of CpGs retained.</li>
</ul>
<h3 id="raises_7">Raises<a class="headerlink" href="#raises_7" title="Permanent link">&para;</a></h3>
<p>ValueError
    If <code>min_samples_per_group</code> is supplied without <code>groups</code> or if group         labels are misaligned.</p>
<hr />
<p><code>filter_min_per_group(M: 'pd.DataFrame', groups: 'pd.Series', min_per_group: 'int' = 5, verbose: 'bool' = True) -&gt; 'pd.DataFrame'</code></p>
<p>Retain only CpGs with at least <code>min_per_group</code> non-missing     values in one experimental group.</p>
<p>Useful for ensuring sufficient representation before differential analysis.</p>
<h3 id="parameters_25">Parameters<a class="headerlink" href="#parameters_25" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix (CpGs  samples) possibly containing NaN.</li>
<li><strong>groups</strong>: <code>pd.Series</code>
    Group membership for each sample (index must match <code>M.columns</code>).</li>
<li><strong>min_per_group</strong>: <code>int, default 5</code>
    Minimum number of observed (non-missing) values required per group.</li>
<li><strong>verbose</strong>: <code>bool, default True</code>
    Log a concise summary of filtering results.</li>
</ul>
<h3 id="returns_23">Returns<a class="headerlink" href="#returns_23" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Filtered matrix containing only qualifying CpGs.</p>
<h3 id="raises_8">Raises<a class="headerlink" href="#raises_8" title="Permanent link">&para;</a></h3>
<p>ValueError
    If group labels do not cover all samples in <code>M</code>.</p>
<hr />
<p><code>impute_missing_values(M: 'pd.DataFrame', method: 'str' = 'mean', k: 'int' = 5, use_sample_knn: 'bool' = False) -&gt; 'pd.DataFrame'</code></p>
<p>Impute missing methylation values using row-wise statistics or K-nearest neighbours.</p>
<h3 id="parameters_26">Parameters<a class="headerlink" href="#parameters_26" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    CpG  sample matrix with possible NaN entries.</li>
<li><strong>method</strong>: <code>{"mean", "median", "knn"}, default "mean"</code>
    Imputation strategy.</li>
<li><strong>k</strong>: <code>int, default 5</code>
    Number of neighbours used for KNN imputation.</li>
<li><strong>use_sample_knn</strong>: <code>bool, default False</code>
    If True, KNN is performed across samples (samples  CpGs) rather         than across CpGs.
    Dramatically faster for typical array datasets.</li>
</ul>
<h3 id="returns_24">Returns<a class="headerlink" href="#returns_24" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Imputed matrix with the same index/columns as input.</p>
<h3 id="notes_7">Notes<a class="headerlink" href="#notes_7" title="Permanent link">&para;</a></h3>
<ul>
<li><code>mean</code> and <code>median</code> are applied row-wise (per CpG).</li>
<li><code>knn</code> uses scikit-learn's <code>KNNImputer</code> with distance weighting.</li>
<li>No imputation is performed if the matrix contains no missing values.</li>
</ul>
<h3 id="validation">Validation<a class="headerlink" href="#validation" title="Permanent link">&para;</a></h3>
<p>Input validation and integrity checks for differential methylation analysis.</p>
<ul>
<li>This module implements rigorous pre-analysis validation of data matrices, experimental designs, contrasts, sample alignment, and system resources. - It ensures statistical estimability, correct pairing structure, and sufficient memory before launching computationally intensive analyses, preventing silent failures and ambiguous results.</li>
</ul>
<h3 id="features_5">Features<a class="headerlink" href="#features_5" title="Permanent link">&para;</a></h3>
<ul>
<li>Accurate memory footprint estimation with conservative peak usage prediction and automatic MemoryError on insufficient RAM</li>
<li>Strict validation of two-group design vectors with automatic construction of intercept + indicator design matrix</li>
<li>Comprehensive contrast validation including shape, zero-vector detection, and QR-based estimability checking</li>
<li>Flexible string contrast syntax ("treatment-control") for simple two-column designs</li>
<li>Thorough alignment verification between methylation data columns, design matrix rows, and optional sample identifiers</li>
<li>Robust paired-sample validation ensuring exactly two observations per subject and balanced group representation within pairs</li>
<li>Clear, actionable error messages and logging for rapid debugging in production pipelines</li>
</ul>
<hr />
<p><code>check_analysis_memory(M: 'pd.DataFrame', warn_threshold_gb: 'float' = 8.0)</code></p>
<p>Conservatively estimate RAM requirements for a full-matrix differential analysis.</p>
<ul>
<li>Calculates current data footprint, projects peak usage (approximately 4 times     input size), and compares against available system memory.</li>
<li>Raises MemoryError early if danger is high, forcing use of chunked mode.</li>
</ul>
<h3 id="parameters_27">Parameters<a class="headerlink" href="#parameters_27" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix (CpGs  samples).</li>
<li><strong>warn_threshold_gb</strong>: <code>float, default 8.0</code>
    Issue a warning if estimated peak exceeds this value.</li>
</ul>
<h3 id="returns_25">Returns<a class="headerlink" href="#returns_25" title="Permanent link">&para;</a></h3>
<p>dict
    Keys: <code>data_gb</code>, <code>peak_gb</code>, <code>available_gb</code>.</p>
<h3 id="raises_9">Raises<a class="headerlink" href="#raises_9" title="Permanent link">&para;</a></h3>
<p>MemoryError
    If projected peak consumption exceeds ~80% of available RAM. In such cases         <code>fit_differential_chunked</code> should be used instead.</p>
<hr />
<p><code>validate_alignment(data: 'pd.DataFrame', design_matrix: 'np.ndarray', sample_names: 'Optional[Sequence]' = None, paired_ids: 'Optional[Sequence]' = None, group_col_idx: 'int' = 1) -&gt; 'Tuple[pd.Index, np.ndarray, Optional[np.ndarray]]'</code></p>
<p>Ensure perfect alignment between methylation data, design matrix, and optional     pairing information.</p>
<p>Verifies sample counts, column ordering, duplicate names, and (if supplied)     paired-subject structure.</p>
<h3 id="parameters_28">Parameters<a class="headerlink" href="#parameters_28" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>data</strong>: <code>pd.DataFrame</code>
    Methylation matrix with samples in columns.</li>
<li><strong>design_matrix</strong>: <code>np.ndarray</code>
    Corresponding design matrix.</li>
<li><strong>sample_names</strong>: <code>Sequence or None</code>
    Explicit ordered list of sample identifiers (required only         if column names differ).</li>
<li><strong>paired_ids</strong>: <code>Sequence or None</code>
    Subject/block identifiers for paired designs; each ID must appear exactly twice.</li>
<li><strong>group_col_idx</strong>: <code>int, default 1</code>
    Column in design_matrix encoding group membership (used to verify         balanced pairing).</li>
</ul>
<h3 id="returns_26">Returns<a class="headerlink" href="#returns_26" title="Permanent link">&para;</a></h3>
<p>tuple
    (ordered_data_columns, validated_design_matrix, validated_paired_array_or_None)</p>
<h3 id="raises_10">Raises<a class="headerlink" href="#raises_10" title="Permanent link">&para;</a></h3>
<p>ValueError
    On any misalignment, duplicate column names, or invalid pairing structure         (e.g., a subject with both samples in the same group).</p>
<hr />
<p><code>validate_contrast(design_matrix: 'np.ndarray', contrast: 'Union[np.ndarray, Sequence[float], str]') -&gt; 'np.ndarray'</code></p>
<p>Validate and normalise a contrast against a given design matrix.</p>
<ul>
<li><strong>Supports</strong>: <code>- Numeric vectors/matrices</code></li>
<li>Simple string syntax <code>"treatment-control"</code> (only for two-column     intercept + group designs)</li>
</ul>
<p>Checks dimensionality, non-zero status, and estimability via QR decomposition.</p>
<h3 id="parameters_29">Parameters<a class="headerlink" href="#parameters_29" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>design_matrix</strong>: <code>np.ndarray</code>
    Full design matrix (n_samples  n_coefficients).</li>
<li><strong>contrast</strong>: <code>np.ndarray, Sequence[float], or str</code>
    Contrast specification.</li>
</ul>
<h3 id="returns_27">Returns<a class="headerlink" href="#returns_27" title="Permanent link">&para;</a></h3>
<p>np.ndarray
    Validated contrast vector (float64.</p>
<h3 id="raises_11">Raises<a class="headerlink" href="#raises_11" title="Permanent link">&para;</a></h3>
<p>ValueError
    On shape mismatch, zero contrast, or non-estimable contrast.</p>
<hr />
<p><code>validate_design(design: 'Sequence') -&gt; 'pd.DataFrame'</code></p>
<p>Convert a simple two-group label vector into a proper design matrix.</p>
<p>Accepts lists, tuples, arrays, or pandas Series containing exactly two     distinct group labels and returns a two-column matrix:</p>
<ul>
<li>Column 0: intercept (all 1s)</li>
<li>Column 1: group indicator (0 = reference group, 1 = alternative)</li>
</ul>
<h3 id="parameters_30">Parameters<a class="headerlink" href="#parameters_30" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>design</strong>: <code>Sequence</code>
    Vector of group assignments (length  2).</li>
</ul>
<h3 id="returns_28">Returns<a class="headerlink" href="#returns_28" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    n_sample  2 design matrix acceptable for <code>fit_differential</code>.</p>
<h3 id="raises_12">Raises<a class="headerlink" href="#raises_12" title="Permanent link">&para;</a></h3>
<p>TypeError
    If input is not a supported sequence type.
ValueError
    If fewer than two samples, NaNs present, or more/less than two unique groups.</p>
<hr />
<h2 id="downstream-analysis">Downstream Analysis<a class="headerlink" href="#downstream-analysis" title="Permanent link">&para;</a></h2>
<h3 id="annotation">Annotation<a class="headerlink" href="#annotation" title="Permanent link">&para;</a></h3>
<p>Annotation and functional interpretation utilities for DNA methylation analysis.</p>
<p>This module delivers a complete suite of tools for translating CpG- and region-level differential methylation results into biological context. It enables rapid gene annotation, pathway- and gene-set enrichment analysis, computation of pathway-level methylation activity, correlation with gene expression, and reliable coordinate liftover across genome assemblies  all implemented with performance and robustness suitable for genome-wide studies.</p>
<h3 id="features_6">Features<a class="headerlink" href="#features_6" title="Permanent link">&para;</a></h3>
<ul>
<li>Ultra-fast nearest-gene annotation using IntervalTree with graceful fallback to distance-based mapping</li>
<li>Fishers exact test gene-set enrichment with automatic background correction, size filtering, and FDR adjustment</li>
<li>Flexible pathway-level methylation scoring (mean/median/sum) from user-provided gene-to-pathway mappings</li>
<li>Sample-aware Pearson/Spearman correlation between CpG methylation and gene expression with robust overlap handling</li>
<li>High-accuracy genomic liftover via pyliftover (hg19  hg38 and other builds) with per-region success tracking</li>
<li>Comprehensive handling of missing dependencies, empty inputs, and chromosome naming conventions</li>
<li>Preservation of original indices and seamless integration with DMeth result tables</li>
</ul>
<hr />
<p><code>adjust_pvalues(pvals: Union[pandas.core.series.Series, ndarray, List[float]], method: str = 'fdr_bh') -&gt; pandas.core.series.Series</code></p>
<p>Apply multiple-testing correction to raw p-values using statsmodels.</p>
<p>Handles NaN values robustly (treated as non-significant) and preserves the     original index when input is a pandas Series.</p>
<h3 id="parameters_31">Parameters<a class="headerlink" href="#parameters_31" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>pvals</strong>: <code>array-like</code>
    Raw p-values (0  p  1).</li>
<li><strong>method</strong>: <code>str, default "fdr_bh"</code>
    Correction method passed to <code>statsmodels.stats.multitest.multipletests</code>.
    Supported: <code>"bonferroni"</code>, <code>"holm"</code>, <code>"fdr_bh"</code>, <code>"fdr_by"</code>,         <code>"sidak"</code>, <code>"none"</code>.</li>
</ul>
<h3 id="returns_29">Returns<a class="headerlink" href="#returns_29" title="Permanent link">&para;</a></h3>
<p>pd.Series
    Adjusted p-values with identical index/order as input.</p>
<hr />
<p><code>annotate_dms_with_genes(dms: 'pd.DataFrame', genes: 'pd.DataFrame', cpg_chr_col: 'str' = 'chr', cpg_pos_col: 'str' = 'pos', gene_chr_col: 'str' = 'chr', gene_start_col: 'str' = 'start', gene_end_col: 'str' = 'end', gene_name_col: 'str' = 'gene_symbol', max_distance: 'int' = 5000) -&gt; 'pd.DataFrame'</code></p>
<p>Assign nearest gene(s) to each differentially methylated site     (DMS) using genomic coordinates.</p>
<p>Uses IntervalTree for ultra-fast exact overlap queries when available;     automatically falls back to distance-based nearest-gene mapping if the     dependency is missing.</p>
<h3 id="parameters_32">Parameters<a class="headerlink" href="#parameters_32" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>dms</strong>: <code>pd.DataFrame</code>
    Table of differentially methylated CpGs with chromosome and position columns.</li>
<li><strong>genes</strong>: <code>pd.DataFrame</code>
    Gene annotation table containing chromosome, start, end, and gene symbol.
cpg_chr_col, cpg_pos_col : str
    Column names in <code>dms</code> for chromosome and position (default: "chr", "pos").
gene_chr_col, gene_start_col, gene_end_col, gene_name_col : str
    Corresponding column names in <code>genes</code> (default: "chr", "start",         "end", "gene_symbol").</li>
<li><strong>max_distance</strong>: <code>int, default 5000</code>
    Maximum distance (bp) considered for nearest-gene assignment when         no overlap exists.</li>
</ul>
<h3 id="returns_30">Returns<a class="headerlink" href="#returns_30" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Original <code>dms</code> with an added <code>nearest_gene</code> column containing         comma-separated gene symbols (or NaN if none found).</p>
<h3 id="notes_8">Notes<a class="headerlink" href="#notes_8" title="Permanent link">&para;</a></h3>
<ul>
<li>Handles mixed chromosome naming conventions via internal normalization.</li>
<li>Multiple overlapping genes are reported (e.g., bidirectional promoters).</li>
</ul>
<hr />
<p><code>correlate_methylation_expression(beta: 'pd.DataFrame', expression: 'pd.DataFrame', gene_map: 'Optional[Dict[str, str]]' = None, method: 'str' = 'pearson', override_index_alignment: 'bool' = False) -&gt; 'pd.DataFrame'</code></p>
<p>Compute sample-wise correlation between CpG methylation and gene expression.</p>
<p>Supports both one-to-one (same index) and many-to-one (custom CpGgene     mapping) scenarios.</p>
<h3 id="parameters_33">Parameters<a class="headerlink" href="#parameters_33" title="Permanent link">&para;</a></h3>
<p>beta, expression : pd.DataFrame
    Methylation (CpGs  samples) and expression (genes  samples) matrices.
- <strong>gene_map</strong>: <code>dict or None</code>
    Explicit mapping from CpG ID  gene symbol (required for cis-analysis).
- <strong>method</strong>: <code>{"pearson", "spearman"}, default "pearson"</code>
    Correlation coefficient to compute.
- <strong>override_index_alignment</strong>: <code>bool, default False</code>
    Proceed even with &lt;2 overlapping samples (useful for exploratory checks).</p>
<h3 id="returns_31">Returns<a class="headerlink" href="#returns_31" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Correlation results with columns <code>r</code> and <code>pval</code> (indexed by CpG         and optionally gene).</p>
<hr />
<p><code>gene_set_enrichment(gene_list: 'List[str]', background: 'Optional[List[str]]' = None, gene_sets: 'Optional[Dict[str, List[str]]]' = None, method: 'str' = 'fisher', pval_cutoff: 'float' = 0.05, min_set_size: 'int' = 5, max_set_size: 'int' = 500) -&gt; 'pd.DataFrame'</code></p>
<p>Perform over-representation analysis (Fishers exact test) on a list     of genes against predefined gene sets (e.g., GO, KEGG, Reactome).</p>
<p>Automatically filters gene sets by size and applies FDR correction.</p>
<h3 id="parameters_34">Parameters<a class="headerlink" href="#parameters_34" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>gene_list</strong>: <code>list[str]</code>
    Genes of interest (e.g., nearest genes of significant DMS/DMRs).</li>
<li><strong>background</strong>: <code>list[str] or None</code>
    Background gene universe. Defaults to union of all genes in <code>gene_sets</code>.</li>
<li><strong>gene_sets</strong>: <code>dict[str, list[str]]</code>
    Mapping from pathway/term name to member genes.
min_set_size, max_set_size : int
    Exclude overly small or large gene sets (default 5500).</li>
<li><strong>pval_cutoff</strong>: <code>float, default 0.05</code>
    Return only terms with adjusted p-value  this threshold.</li>
</ul>
<h3 id="returns_32">Returns<a class="headerlink" href="#returns_32" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Enriched terms with columns: <code>term</code>, <code>pvalue</code>, <code>padj</code>,         <code>oddsratio</code>, <code>overlap</code>, <code>set_size</code>.</p>
<hr />
<p><code>liftover_coordinates(regions: 'pd.DataFrame', from_build: 'str' = 'hg19', to_build: 'str' = 'hg38', chr_col: 'str' = 'chr', start_col: 'str' = 'start', end_col: 'str' = 'end') -&gt; 'pd.DataFrame'</code></p>
<p>Convert genomic regions between genome assemblies using pyliftover     (e.g., hg19  hg38).</p>
<p>Handles both single-base and interval coordinates with per-row success reporting.</p>
<h3 id="parameters_35">Parameters<a class="headerlink" href="#parameters_35" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>regions</strong>: <code>pd.DataFrame</code>
    Table containing chromosome, start, and end columns.
from_build, to_build : str
    Source and target genome builds (default: hg19  hg38).
chr_col, start_col, end_col : str
    Column names for genomic coordinates.</li>
</ul>
<h3 id="returns_33">Returns<a class="headerlink" href="#returns_33" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Original table augmented with:
    <code>lifted_chr</code>, <code>lifted_start</code>, <code>lifted_end</code>
    <code>lifted</code> (boolean indicating successful conversion)</p>
<h3 id="raises_13">Raises<a class="headerlink" href="#raises_13" title="Permanent link">&para;</a></h3>
<p>RuntimeError
    If <code>pyliftover</code> is not installed.</p>
<hr />
<p><code>pathway_methylation_scores(beta: 'pd.DataFrame', annotation: 'pd.DataFrame', pathway_db: 'Dict[str, List[str]]', method: 'str' = 'mean') -&gt; 'pd.DataFrame'</code></p>
<p>Collapse CpG-level beta values into pathway-level methylation scores per sample.</p>
<p>Useful for downstream pathway-activity modeling or visualization.</p>
<h3 id="parameters_36">Parameters<a class="headerlink" href="#parameters_36" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>beta</strong>: <code>pd.DataFrame</code>
    Beta-value matrix (CpGs  samples).</li>
<li><strong>annotation</strong>: <code>pd.DataFrame</code>
    Mapping from CpG identifiers to gene symbols (index  gene).</li>
<li><strong>pathway_db</strong>: <code>dict[str, list[str]]</code>
    Dictionary of pathways  list of associated genes.</li>
<li><strong>method</strong>: <code>{"mean", "median", "sum"}, default "mean"</code>
    Aggregation function applied across CpGs belonging to each pathway.</li>
</ul>
<h3 id="returns_34">Returns<a class="headerlink" href="#returns_34" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Pathways  samples matrix of aggregated methylation scores.</p>
<h3 id="deconvolution">Deconvolution<a class="headerlink" href="#deconvolution" title="Permanent link">&para;</a></h3>
<p>Cell-type deconvolution utilities for bulk DNA methylation data.</p>
<p>This module provides fast, reference-based estimation of cell-type proportions from epigenome-wide methylation profiles using constrained non-negative least squares (NNLS). It supports integration of published reference matrices (e.g., blood, brain, tumor microenvironment) and enables accurate correction for cellular heterogeneity in differential methylation studies.</p>
<h3 id="features_7">Features<a class="headerlink" href="#features_7" title="Permanent link">&para;</a></h3>
<ul>
<li>High-performance NNLS deconvolution with automatic feature alignment between bulk and reference</li>
<li>Optional parallel processing via joblib for large cohorts (hundreds to thousands of samples)</li>
<li>Robust normalization of estimated proportions to sum-to-1 per sample</li>
<li>Graceful handling of missing joblib dependency with automatic fallback to serial execution</li>
<li>Clear error reporting when no overlapping CpGs exist between dataset and reference</li>
<li>Direct compatibility with standard beta-value matrices (CpGs  samples) used throughout DMeth)</li>
<li>Extensible design for future reference-free (e.g., EpiDISH HEpiDISH, MeDeCom) or projective methods</li>
</ul>
<hr />
<p><code>estimate_cell_composition(beta: DataFrame, ref_profiles: DataFrame, method: str = 'nnls', n_jobs: int = 1) -&gt; DataFrame</code></p>
<p>Perform reference-based cell-type deconvolution of bulk DNA methylation     data using constrained non-negative least squares (NNLS).</p>
<p>Estimates the relative proportions of predefined cell types in each     sample by solving the linear mixture model:
    _bulk  R  w</p>
<p>where R contains cell-type-specific methylation reference profiles     and w  0 are the unknown proportions.</p>
<h3 id="parameters_37">Parameters<a class="headerlink" href="#parameters_37" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>beta</strong>: <code>pd.DataFrame</code>
    Bulk beta-value matrix with CpGs as rows (index) and samples as columns.
    Values must be in [0, 1]; missing values are not supported.</li>
<li><strong>ref_profiles</strong>: <code>pd.DataFrame</code>
    Reference matrix of pure cell-type methylation profiles.
    Must have the same CpG index as <code>beta</code> (or a subset thereof) and cell         types as columns.</li>
<li><strong>method</strong>: <code>str, default "nnls"</code>
    Deconvolution algorithm. Currently only <code>"nnls"</code> (scipy.optimize.nnls)         is implemented.</li>
<li><strong>n_jobs</strong>: <code>int, default 1</code>
    Number of parallel processes for sample-wise deconvolution.
    Requires <code>joblib</code>; automatically falls back to serial execution if         unavailable or set to 1.</li>
</ul>
<h3 id="returns_35">Returns<a class="headerlink" href="#returns_35" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Estimated cell-type proportions with:
    rows = samples (same order and names as <code>beta.columns</code>)
    columns = cell types (same as <code>ref_profiles.columns</code>)
    values constrained to 0 and normalized to sum to 1 per sample</p>
<h3 id="raises_14">Raises<a class="headerlink" href="#raises_14" title="Permanent link">&para;</a></h3>
<p>ValueError
    If no overlapping CpG features exist between <code>beta</code> and <code>ref_profiles</code>.</p>
<h3 id="notes_9">Notes<a class="headerlink" href="#notes_9" title="Permanent link">&para;</a></h3>
<ul>
<li>Automatic feature alignment: only CpGs present in both matrices are used.</li>
<li>Proportions are forcibly normalized to sum to 1 (with protection against     division by zero).</li>
<li>Highly efficient for typical blood, brain, or tumor microenvironment references     (e.g., FlowSorted.Blood.EPIC, Houseman, etc.).</li>
<li>Designed for seamless integration with DMeth preprocessing pipelines     (beta-value input expected).</li>
</ul>
<h3 id="downstream-stats">Downstream Stats<a class="headerlink" href="#downstream-stats" title="Permanent link">&para;</a></h3>
<p>Core statistical utilities for DNA methylation analysis downstream of differential testing.</p>
<ul>
<li>This module provides a comprehensive suite of high-performance, publication-grade functions for p-value adjustment, effect-size computation, delta-beta calculation, reproducible DMR calling via sliding-window clustering, regional summarization, and cross-dataset reproducibility assessment.</li>
<li>Designed for seamless integration into DMeth pipelines, all functions operate efficiently on large epigenome-wide datasets while preserving genomic coordinates and CpG identifiers.</li>
</ul>
<h3 id="features_8">Features<a class="headerlink" href="#features_8" title="Permanent link">&para;</a></h3>
<ul>
<li>Multiple p-value correction methods via statsmodels with robust NaN handling</li>
<li>Stouffers Z-score method for meta-analysis across studies or batches</li>
<li>Cohens d and Hedges g effect size estimation with proper small-sample correction</li>
<li>Flexible delta-beta computation with optional absolute values and index alignment</li>
<li>Threshold-based filtering of DMS results supporting logFC, delta-beta, and directional constraints</li>
<li>Fast, vectorized sliding-window DMR discovery with configurable gap merging and minimum CpG requirements</li>
<li>Comprehensive DMR summarization and cross-dataset reproducibility metrics (Jaccard, concordance, Spearman correlation)</li>
<li>Full preservation of CpG identifiers and genomic coordinates throughout all operations</li>
</ul>
<hr />
<p><code>adjust_pvalues(pvals: Union[pandas.core.series.Series, ndarray, List[float]], method: str = 'fdr_bh') -&gt; pandas.core.series.Series</code></p>
<p>Apply multiple-testing correction to raw p-values using statsmodels.</p>
<p>Handles NaN values robustly (treated as non-significant) and preserves the     original index when input is a pandas Series.</p>
<h3 id="parameters_38">Parameters<a class="headerlink" href="#parameters_38" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>pvals</strong>: <code>array-like</code>
    Raw p-values (0  p  1).</li>
<li><strong>method</strong>: <code>str, default "fdr_bh"</code>
    Correction method passed to <code>statsmodels.stats.multitest.multipletests</code>.
    Supported: <code>"bonferroni"</code>, <code>"holm"</code>, <code>"fdr_bh"</code>, <code>"fdr_by"</code>,         <code>"sidak"</code>, <code>"none"</code>.</li>
</ul>
<h3 id="returns_36">Returns<a class="headerlink" href="#returns_36" title="Permanent link">&para;</a></h3>
<p>pd.Series
    Adjusted p-values with identical index/order as input.</p>
<hr />
<p><code>compute_delta_beta(mean_beta_group1: Union[pandas.core.series.Series, ndarray], mean_beta_group2: Union[pandas.core.series.Series, ndarray], as_abs: bool = False) -&gt; pandas.core.series.Series</code></p>
<p>Calculate per-CpG difference in mean beta values between two groups.</p>
<p>Automatically aligns inputs by index; supports absolute differences.</p>
<h3 id="parameters_39">Parameters<a class="headerlink" href="#parameters_39" title="Permanent link">&para;</a></h3>
<p>mean_beta_group1, mean_beta_group2 : Series-like
    Group-wise mean beta values.
- <strong>as_abs</strong>: <code>bool, default False</code>
    Return absolute delta-beta if True.</p>
<h3 id="returns_37">Returns<a class="headerlink" href="#returns_37" title="Permanent link">&para;</a></h3>
<p>pd.Series
    Delta-beta values (positive = group1 &gt; group2).</p>
<hr />
<p><code>compute_dms_reproducibility(res1: DataFrame, res2: DataFrame, id_col: Optional[str] = None, effect_col: str = 'logFC', pval_col: str = 'padj', pval_thresh: float = 0.05) -&gt; Dict[str, Any]</code></p>
<p>Quantify reproducibility of differential methylation signals across     two independent analyses or cohorts.</p>
<p>Metrics include:</p>
<ul>
<li>Total/overlap feature counts</li>
<li>Jaccard index</li>
<li>Number of overlapping significant CpGs</li>
<li>Directional concordance</li>
<li>Spearman correlation of effect sizes</li>
</ul>
<h3 id="parameters_40">Parameters<a class="headerlink" href="#parameters_40" title="Permanent link">&para;</a></h3>
<p>res1, res2 : pd.DataFrame
    Two differential result tables (same CpG index preferred).
effect_col, pval_col : str
    Columns for effect size and adjusted p-value.
- <strong>pval_thresh</strong>: <code>float, default 0.05</code></p>
<h3 id="returns_38">Returns<a class="headerlink" href="#returns_38" title="Permanent link">&para;</a></h3>
<p>dict
    Comprehensive reproducibility statistics.</p>
<hr />
<p><code>compute_effect_size(beta_group1: DataFrame, beta_group2: DataFrame, method: str = 'cohens_d') -&gt; pandas.core.series.Series</code></p>
<p>Compute standardized effect size (Cohens d or Hedges g) for each CpG     between two groups.</p>
<p>Uses pooled standard deviation with proper small-sample correction for Hedges g.</p>
<h3 id="parameters_41">Parameters<a class="headerlink" href="#parameters_41" title="Permanent link">&para;</a></h3>
<p>beta_group1, beta_group2 : pd.DataFrame
    Beta matrices (CpGs  samples) for each group.
- <strong>method</strong>: <code>{"cohens_d", "hedges_g"}, default "cohens_d"</code></p>
<h3 id="returns_39">Returns<a class="headerlink" href="#returns_39" title="Permanent link">&para;</a></h3>
<p>pd.Series
    Effect size per CpG (positive = higher in group1).</p>
<hr />
<p><code>filter_dms(res: DataFrame, lfc_col: str = 'logFC', pval_col: str = 'padj', delta_beta_col: Optional[str] = None, pval_thresh: float = 0.05, lfc_thresh: float = 0.0, delta_beta_thresh: Optional[float] = None, direction: Optional[str] = None) -&gt; DataFrame</code></p>
<p>Apply multi-criterion filtering to differential methylation results.</p>
<ul>
<li><strong>Supports</strong>: <code>- Adjusted p-value threshold</code></li>
<li>Minimum |logFC|</li>
<li>Minimum ||</li>
<li>Directional filtering (hyper or hypo)</li>
</ul>
<h3 id="parameters_42">Parameters<a class="headerlink" href="#parameters_42" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>res</strong>: <code>pd.DataFrame</code>
    Differential results table.
lfc_col, pval_col, delta_beta_col : str
    Column names (defaults: "logFC", "padj", optional).
pval_thresh, lfc_thresh, delta_beta_thresh
    Numeric thresholds.</li>
<li><strong>direction</strong>: <code>{"hyper", "hypo", None}</code></li>
</ul>
<h3 id="returns_40">Returns<a class="headerlink" href="#returns_40" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Subset of rows passing all specified criteria.</p>
<hr />
<p><code>find_dmrs_by_sliding_window(dms: DataFrame, annotation: DataFrame, chr_col: str = 'chr', pos_col: str = 'pos', pval_col: str = 'padj', pval_thresh: float = 0.05, delta_beta_col: Optional[str] = 'delta_beta', max_gap: int = 500, min_cpgs: int = 3, merge_distance: Optional[int] = None, use_intervaltree: bool = True) -&gt; DataFrame</code></p>
<p>Identify differentially methylated regions (DMRs) by clustering spatially     proximate significant CpGs.</p>
<p>Uses a fast sliding-window/gap-merging approach per chromosome.</p>
<h3 id="parameters_43">Parameters<a class="headerlink" href="#parameters_43" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>dms</strong>: <code>pd.DataFrame</code>
    Significant DMS results (after filtering).</li>
<li><strong>annotation</strong>: <code>pd.DataFrame</code>
    CpG annotation with <code>chr</code> and <code>pos</code> columns.</li>
<li><strong>max_gap</strong>: <code>int, default 500</code>
    Maximum distance (bp) to bridge adjacent significant CpGs.</li>
<li><strong>min_cpgs</strong>: <code>int, default 3</code>
    Minimum number of significant CpGs required to call a DMR.</li>
<li><strong>merge_distance</strong>: <code>int or None</code>
    If set, merge DMRs closer than this distance.</li>
</ul>
<h3 id="returns_41">Returns<a class="headerlink" href="#returns_41" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    One row per DMR with columns:
    <code>chr</code>, <code>start</code>, <code>end</code>, <code>n_cpgs</code>, <code>mean_delta_beta</code>,         <code>mean_logFC</code>, <code>min_padj</code>, <code>cpgs</code> (list).</p>
<hr />
<p><code>stouffer_combined_pvalue(pvals)</code></p>
<p>Combine independent p-values across studies or replicates using Stouffers     Z-score method (equal weighting).</p>
<h3 id="parameters_44">Parameters<a class="headerlink" href="#parameters_44" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>pvals</strong>: <code>array-like</code>
    List or array of p-values to combine.</li>
</ul>
<h3 id="returns_42">Returns<a class="headerlink" href="#returns_42" title="Permanent link">&para;</a></h3>
<p>float
    Single combined p-value.</p>
<h3 id="raises_15">Raises<a class="headerlink" href="#raises_15" title="Permanent link">&para;</a></h3>
<p>ValueError
    If input is empty or contains values outside [0, 1].</p>
<hr />
<p><code>summarize_regions(dmrs: DataFrame, summary_cols: Optional[List[str]] = None) -&gt; DataFrame</code></p>
<p>Generate a concise summary table of discovered DMRs.</p>
<p>Reports total count, median length, average CpG density, mean delta-beta,     and strongest significance.</p>
<h3 id="parameters_45">Parameters<a class="headerlink" href="#parameters_45" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>dmrs</strong>: <code>pd.DataFrame</code>
    Output from <code>find_dmrs_by_sliding_window</code>.</li>
<li><strong>summary_cols</strong>: <code>list[str] or None</code>
    Columns to summarize (defaults to key metrics).</li>
</ul>
<h3 id="returns_43">Returns<a class="headerlink" href="#returns_43" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Single-row summary suitable for manuscript tables.</p>
<h3 id="helpers">Helpers<a class="headerlink" href="#helpers" title="Permanent link">&para;</a></h3>
<p>General-purpose utilities for DNA methylation analysis workflows.</p>
<p>This lightweight module contains essential helper functions used across the DMeth package for consistent data handling and summarization. It provides robust chromosome string normalization and flexible group-wise aggregation of beta-value matrices, ensuring compatibility and reproducibility in downstream statistical and annotation steps.</p>
<h3 id="features_9">Features<a class="headerlink" href="#features_9" title="Permanent link">&para;</a></h3>
<ul>
<li>Robust chromosome identifier normalization with automatic 'chr' prefix addition and preservation of X/Y conventions</li>
<li>Safe handling of mixed chromosome formats (e.g., '1', 'chr1', 'CHR1'  'chr1', 'chrX')</li>
<li>Efficient group-wise summarization of per-sample beta values with support for arbitrary aggregation functions</li>
<li>Automatic computation of both mean and variance per group for interpretability and modeling</li>
<li>Graceful handling of missing groups, empty inputs, and non-overlapping sample sets</li>
<li>Index-preserving operations fully compatible with CpG  sample matrix conventions</li>
</ul>
<hr />
<p><code>summarize_groups(beta: DataFrame, groups: pandas.core.series.Series, summary_func: Callable = &lt;function mean at 0x78fd7b058370&gt;) -&gt; DataFrame</code></p>
<p>Compute group-wise summary statistics (mean and variance) across samples     for a beta-value matrix.</p>
<p>Designed for rapid interpretation of methylation levels across     experimental conditions or cell types.</p>
<h3 id="parameters_46">Parameters<a class="headerlink" href="#parameters_46" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>beta</strong>: <code>pd.DataFrame</code>
    Beta-value matrix with CpGs as rows (index) and samples as columns.</li>
<li><strong>groups</strong>: <code>pd.Series</code>
    Sample-to-group mapping. Index must align with <code>beta.columns</code>.</li>
<li><strong>summary_func</strong>: `callable, default <code>np.mean```
    Aggregation function applied per group (e.g.,</code>np.mean<code>,</code>np.median``).</li>
</ul>
<h3 id="returns_44">Returns<a class="headerlink" href="#returns_44" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    New DataFrame with columns:
    <code>mean_{group}</code>  group-wise summary using <code>summary_func</code>
    <code>var_{group}</code>   sample variance within each group (ddof=1)</p>
<div class="codehilite"><pre><span></span><code><span class="n">Missing</span><span class="w"> </span><span class="k">groups</span><span class="w"> </span><span class="k">or</span><span class="w"> </span><span class="k">empty</span><span class="w"> </span><span class="n">intersections</span><span class="w"> </span><span class="n">yield</span><span class="w"> </span><span class="n n-Quoted">``</span><span class="n">NaN</span><span class="n n-Quoted">``</span><span class="w"> </span><span class="k">columns</span><span class="p">.</span>
</code></pre></div>

<h3 id="notes_10">Notes<a class="headerlink" href="#notes_10" title="Permanent link">&para;</a></h3>
<ul>
<li>Fully preserves CpG index.</li>
<li>Handles partial or missing group overlap gracefully.</li>
<li>Ideal for generating input tables for delta-beta calculation,     visualization, or reporting.</li>
</ul>
<h3 id="signature">Signature<a class="headerlink" href="#signature" title="Permanent link">&para;</a></h3>
<p>Biomarker signature discovery and predictive modeling for DNA methylation data.</p>
<p>This module provides a streamlined, production-ready framework for translating differential methylation results into high-performance diagnostic, prognostic, or predictive biomarker panels. It supports feature selection from statistical outputs, cross-validated model training with state-of-the-art algorithms, and rigorous independent validation of classification or regression performance.</p>
<h3 id="features_10">Features<a class="headerlink" href="#features_10" title="Permanent link">&para;</a></h3>
<ul>
<li>Simple yet effective signature selection via top-ranked CpGs by p-value or moderated t-statistic</li>
<li>Flexible predictive modeling with Random Forest and Elastic Net (with built-in hyperparameter tuning via CV)</li>
<li>Automatic task detection and appropriate stratification (classification vs regression)</li>
<li>Comprehensive cross-validation and held-out test set evaluation with standard metrics (AUC, accuracy, R, RMSE)</li>
<li>Full integration with DMeth result tables and sample-level beta matrices (features  samples orientation)</li>
<li>Reproducible training through fixed random seeds and stratified splitting</li>
<li>Extensible architecture for future stability selection, recursive feature elimination, or multi-omics integration</li>
</ul>
<hr />
<p><code>model_dms_for_prediction(beta: 'pd.DataFrame', labels: 'pd.Series', method: 'str' = 'random_forest', n_splits: 'int' = 5, random_state: 'int' = 42, task: 'Optional[str]' = None) -&gt; 'Dict[str, Any]'</code></p>
<p>Train and evaluate a predictive model using DNA methylation signatures.</p>
<p>Automatically detects classification vs regression tasks and applies     appropriate modeling and evaluation strategies.</p>
<h3 id="parameters_47">Parameters<a class="headerlink" href="#parameters_47" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>beta</strong>: <code>pd.DataFrame</code>
    Beta-value matrix with CpGs as rows and samples as columns (features          samples orientation).</li>
<li><strong>labels</strong>: <code>pd.Series</code>
    Target variable aligned with <code>beta.columns</code>.</li>
<li><strong>method</strong>: <code>{"random_forest", "elasticnet"}, default "random_forest"</code>
    Predictive algorithm:
    <code>random_forest</code>: 500 trees with parallel training
    <code>elasticnet</code>: LogisticRegressionCV or ElasticNetCV with built-in         cross-validated regularization</li>
<li><strong>n_splits</strong>: <code>int, default 5</code>
    Number of cross-validation folds for hyperparameter tuning and         performance estimation.</li>
<li><strong>random_state</strong>: <code>int, default 42</code>
    Seed for reproducible splitting and model initialization.</li>
<li><strong>task</strong>: <code>{"classification", "regression"} or None, optional</code>
    Force task type. If <code>None</code>, inferred from label distribution.</li>
</ul>
<h3 id="returns_45">Returns<a class="headerlink" href="#returns_45" title="Permanent link">&para;</a></h3>
<p>dict
    Contains:
    <code>estimator</code>: final fitted model (trained on full data)
    <code>cv_results</code>: detailed cross-validation scores
    <code>test_auc</code> / <code>test_accuracy</code> (classification) or <code>test_rmse</code>         / <code>test_r2</code> (regression) on a stratified 20% held-out set</p>
<hr />
<p><code>select_signature_panel(res: 'pd.DataFrame', method: 'str' = 'top', top_n: 'int' = 100, importance_col: 'str' = 't') -&gt; 'List[str]'</code></p>
<p>Extract a candidate biomarker panel from differential methylation results.</p>
<h3 id="parameters_48">Parameters<a class="headerlink" href="#parameters_48" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>res</strong>: <code>pd.DataFrame</code>
    Differential analysis results table (e.g., output from <code>fit_differential</code>).
    Must contain at least one of <code>pval</code> or the column specified in         <code>importance_col</code>.</li>
<li><strong>method</strong>: <code>str, default "top"</code>
    Feature selection strategy. Currently only <code>"top"</code> is implemented.</li>
<li><strong>top_n</strong>: <code>int, default 100</code>
    Number of top-ranked CpGs to retain.</li>
<li><strong>importance_col</strong>: <code>str, default "t"</code>
    Column used for ranking when <code>pval</code> is unavailable (e.g., moderated         t-statistic).</li>
</ul>
<h3 id="returns_46">Returns<a class="headerlink" href="#returns_46" title="Permanent link">&para;</a></h3>
<p>List[str]
    Ordered list of selected CpG identifiers (row names from <code>res</code>).</p>
<h3 id="notes_11">Notes<a class="headerlink" href="#notes_11" title="Permanent link">&para;</a></h3>
<ul>
<li>Prioritizes <code>pval</code> for ranking if present.</li>
<li>Falls back to descending order of <code>importance_col</code> otherwise.</li>
<li>Future extensions will include stability selection and recursive     feature elimination.</li>
</ul>
<hr />
<p><code>validate_signature(X_train: 'pd.DataFrame', y_train: 'Union[pd.Series, np.ndarray]', X_test: 'pd.DataFrame', y_test: 'Union[pd.Series, np.ndarray]', features: 'Sequence[str]', method: 'str' = 'elasticnet') -&gt; 'Dict[str, Any]'</code></p>
<p>Independent validation of a pre-selected methylation signature on held-out data.</p>
<p>Re-trains the specified model on training data using only the provided     features and reports performance on a separate test set.</p>
<h3 id="parameters_49">Parameters<a class="headerlink" href="#parameters_49" title="Permanent link">&para;</a></h3>
<p>X_train, X_test : pd.DataFrame
    Training and test methylation matrices (samples  features).
y_train, y_test : array-like
    Corresponding labels.
- <strong>features</strong>: <code>Sequence[str]</code>
    Subset of CpGs constituting the signature.
- <strong>method</strong>: <code>str, default "elasticnet"</code>
    Model used for validation (same options as <code>model_dms_for_prediction</code>).</p>
<h3 id="returns_47">Returns<a class="headerlink" href="#returns_47" title="Permanent link">&para;</a></h3>
<p>dict
    Performance metrics:
    Binary classification  <code>auc</code> and <code>accuracy</code>
    Multi-class/regression  <code>accuracy</code> or <code>rmse</code> + <code>r2</code></p>
<h3 id="notes_12">Notes<a class="headerlink" href="#notes_12" title="Permanent link">&para;</a></h3>
<p>Provides an unbiased estimate of clinical/translational performance     when the signature was derived on a separate discovery cohort.</p>
<hr />
<h2 id="configuration">Configuration<a class="headerlink" href="#configuration" title="Permanent link">&para;</a></h2>
<h3 id="config-manager">Config Manager<a class="headerlink" href="#config-manager" title="Permanent link">&para;</a></h3>
<p>Planner configuration manager for DNA methylation studies.</p>
<ul>
<li>Provides a thread-safe singleton (<code>PlannerConfig</code>) that centralizes platforms, experimental designs, cost components, timeline phases, and global settings.</li>
<li>All configurations are strictly validated with Pydantic models and can be atomically loaded from or saved to multiple formats (JSON, YAML, TOML, Python literals, Excel/CSV tables).</li>
</ul>
<h3 id="features_11">Features<a class="headerlink" href="#features_11" title="Permanent link">&para;</a></h3>
<ul>
<li>Global singleton accessible via <code>get_config()</code></li>
<li>Atomic file load/save with automatic merging over built-in defaults</li>
<li>Full Pydantic validation on every load and modification</li>
<li>Excel/CSV sheet-based overrides (Platforms, Designs, Costs, Timeline)</li>
<li>Power-analysis-based sample size calculation with Bonferroni/FDR correction</li>
<li>Timeline and cost estimation helpers</li>
<li>In-place configuration migration facility</li>
</ul>
<hr />
<p><code>CostComponentSchema(*, cost: Annotated[float, Ge(ge=0)], unit: Annotated[str, _PydanticGeneralMetadata(pattern='^(per_sample|per_cpg|fixed)$')], description: Optional[str] = None, optional: bool = False, applies_to: Optional[List[str]] = None) -&gt; None</code></p>
<p><em>No docstring provided.</em></p>
<hr />
<p><code>DesignSchema(*, name: str, description: Optional[str] = None, n_groups: Annotated[int, Ge(ge=1)], paired: bool = False, complexity: Optional[str] = None, min_n_recommended: Annotated[int, Ge(ge=1)], power_adjustment: Annotated[float, Gt(gt=0)], analysis_method: Optional[str] = None, example_uses: Optional[List[str]] = None) -&gt; None</code></p>
<p><em>No docstring provided.</em></p>
<hr />
<p><code>GlobalSettingsSchema(*, contingency_buffer_percent: Annotated[float, Ge(ge=0.0), Le(le=100.0)] = 10.0, default_mcp_method: Annotated[str, _PydanticGeneralMetadata(pattern='^(bonferroni|fdr|none)$')] = 'fdr') -&gt; None</code></p>
<p><em>No docstring provided.</em></p>
<hr />
<p><code>PlannerConfig(config_file: Union[str, pathlib.Path, NoneType] = None)</code></p>
<p>Thread-safe singleton holding the complete planner configuration.</p>
<p>The singleton pattern ensures that only one configuration instance exists
throughout the application lifetime. Access is recommended via the
<code>get_config()</code> function.</p>
<h3 id="parameters_50">Parameters<a class="headerlink" href="#parameters_50" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>config_file</strong>: <code>str or Path, optional</code>
    Configuration file to load immediately after instantiation.</li>
</ul>
<hr />
<p><code>PlannerConfigModel(*, platforms: Dict[str, dmeth.config.config_manager.PlatformSchema], designs: Dict[str, dmeth.config.config_manager.DesignSchema], cost_components: Dict[str, dmeth.config.config_manager.CostComponentSchema], timeline_phases: Dict[str, dmeth.config.config_manager.TimelinePhaseSchema], global_settings: Optional[dmeth.config.config_manager.GlobalSettingsSchema] = GlobalSettingsSchema(contingency_buffer_percent=10.0, default_mcp_method='fdr')) -&gt; None</code></p>
<p><em>No docstring provided.</em></p>
<hr />
<p><code>PlatformSchema(*, name: str, manufacturer: Optional[str] = None, n_cpgs: Annotated[int, Ge(ge=1)], cost_per_sample: Annotated[float, Ge(ge=0)], processing_days: Annotated[int, Ge(ge=0)], dna_required_ng: Optional[float] = None, coverage: Optional[str] = None, release_year: Optional[int] = None, status: Optional[str] = None, recommended: bool = False, notes: Optional[str] = None) -&gt; None</code></p>
<p><em>No docstring provided.</em></p>
<hr />
<p><code>TimelinePhaseSchema(*, name: Optional[str] = None, base_duration_days: Annotated[float, Gt(gt=0)], scaling_factor: Optional[float] = 0.0, batch_adjustment: Optional[float] = 0.0, description: Optional[str] = None, critical: bool = False, optional: bool = False, parallelizable: bool = False) -&gt; None</code></p>
<p><em>No docstring provided.</em></p>
<hr />
<p><code>get_config() -&gt; dmeth.config.config_manager.PlannerConfig</code></p>
<p>Retrieve the global singleton instance of <code>PlannerConfig</code>.</p>
<p>The instance is created on first access if necessary.</p>
<h3 id="returns_48">Returns<a class="headerlink" href="#returns_48" title="Permanent link">&para;</a></h3>
<p>PlannerConfig
    The active configuration object.</p>
<hr />
<p><code>load_file(path: Union[str, pathlib.Path], allow_excel: bool = True) -&gt; None</code></p>
<p>Load and validate a configuration file, merging it with built-in defaults.</p>
<p>Supported formats are JSON, YAML, TOML, Python literal dictionaries,
and Excel/CSV table files. The loaded data is deeply merged over the
defaults and then validated against the Pydantic schema.</p>
<h3 id="parameters_51">Parameters<a class="headerlink" href="#parameters_51" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>path</strong>: <code>str or Path</code>
    Path to the configuration file.</li>
<li><strong>allow_excel</strong>: <code>bool, default True</code>
    Permit loading from Excel (.xlsx, .xls) or CSV files.</li>
</ul>
<h3 id="raises_16">Raises<a class="headerlink" href="#raises_16" title="Permanent link">&para;</a></h3>
<p>FileNotFoundError
    If the specified file does not exist.
PermissionError
    If the file cannot be read.
ValueError
    If the file format is unsupported or malformed.
ValidationError
    If the merged configuration fails schema validation.</p>
<hr />
<p><code>load_from_table_file(path: pathlib.Path) -&gt; Dict[str, Any]</code></p>
<p>Parse configuration fragments from an Excel workbook or CSV file.</p>
<p>Recognised sheets/tables are:</p>
<ul>
<li>Platforms</li>
<li>Designs</li>
<li>Costs (mapped to <code>cost_components</code>)</li>
<li>Timeline (mapped to <code>timeline_phases</code>)</li>
</ul>
<p>The function returns a dictionary containing only the sections that
could be successfully interpreted.</p>
<h3 id="parameters_52">Parameters<a class="headerlink" href="#parameters_52" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>path</strong>: <code>Path</code>
    Path to the Excel or CSV file.</li>
</ul>
<h3 id="returns_49">Returns<a class="headerlink" href="#returns_49" title="Permanent link">&para;</a></h3>
<p>dict
    Partial configuration with keys <code>platforms</code>, <code>designs</code>,
    <code>cost_components</code>, and/or <code>timeline_phases</code>.</p>
<h3 id="raises_17">Raises<a class="headerlink" href="#raises_17" title="Permanent link">&para;</a></h3>
<p>ValueError
    If no recognised configuration tables are found.</p>
<hr />
<p><code>migrate(migrator_callable) -&gt; None</code></p>
<p>Apply an in-place migration to the active configuration.</p>
<p>The provided callable receives a copy of the current raw configuration
and must return an updated dictionary. The result is merged, validated,
and activated.</p>
<h3 id="parameters_53">Parameters<a class="headerlink" href="#parameters_53" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>migrator_callable</strong>: <code>callable</code>
    Function <code>raw_config -&gt; updated_config</code>.</li>
</ul>
<hr />
<p><code>reload() -&gt; None</code></p>
<p>Reload the currently-loaded configuration file.</p>
<p>If no file was loaded earlier, this re-validates the in-memory config.</p>
<hr />
<p><code>reset_config()</code></p>
<p>Destroy the current singleton instance (primarily for testing).</p>
<hr />
<p><code>save_file(path: Union[str, pathlib.Path], fmt: Optional[str] = None) -&gt; None</code></p>
<p>Atomically save the current configuration to a file.</p>
<p>The format is inferred from the file extension unless <code>fmt</code> is provided.</p>
<p>Supported formats: JSON (default), YAML, TOML.</p>
<h3 id="parameters_54">Parameters<a class="headerlink" href="#parameters_54" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>path</strong>: <code>str or Path</code>
    Destination file path.</li>
<li><strong>fmt</strong>: <code>str, optional</code>
    Explicit format (<code>json</code>, <code>yaml</code>, <code>toml</code>).</li>
</ul>
<h3 id="raises_18">Raises<a class="headerlink" href="#raises_18" title="Permanent link">&para;</a></h3>
<p>ValueError
    If the requested format is not supported.
RuntimeError
    If a required library (PyYAML or toml) is missing.</p>
<hr />
<h2 id="inputoutput">Input/Output<a class="headerlink" href="#inputoutput" title="Permanent link">&para;</a></h2>
<h3 id="data-utils">Data Utils<a class="headerlink" href="#data-utils" title="Permanent link">&para;</a></h3>
<p>Core data structures and validation utilities for the DMeth pipeline.</p>
<p>This module defines the central ProcessedData container used throughout the package
and provides helper functions for ensuring data integrity (index alignment and type safety).</p>
<h2 id="key-components">Key Components<a class="headerlink" href="#key-components" title="Permanent link">&para;</a></h2>
<p>ProcessedData</p>
<ul>
<li>
<p>The standard container holding aligned methylation matrix, sample metadata,
and probe annotation. All analysis functions expect this object.</p>
</li>
<li>
<p>Validation helpers:</p>
<ul>
<li><code>_ensure_index_alignment(M, pheno, ann)</code>
    Checks that samples and probes are consistently indexed across components.</li>
<li><code>_ensure_index_strings(df)</code>
    Safely converts DataFrame indices, but not columns, to strings (required         for HDF5/h5py compatibility).</li>
</ul>
</li>
</ul>
<p>All ProcessedData objects are automatically validated and string-indexed upon creation.</p>
<hr />
<p><code>ProcessedData(M: DataFrame, pheno: DataFrame, ann: DataFrame, meta: Dict[str, Any] = &lt;factory&gt;) -&gt; None</code></p>
<p>Central container for aligned methylation data used throughout the DMeth pipeline.</p>
<p>Guarantees that:</p>
<ul>
<li>All components (methylation matrix, sample metadata, probe annotation)     share consistent string indices.</li>
<li>Sample and probe alignment is validated on construction.</li>
<li>A <code>meta</code> dictionary tracks processing history (normalisation,     batch correction, QC metrics, etc.).</li>
</ul>
<p>All downstream functions expect or return instances of this class.</p>
<h3 id="parameters_55">Parameters<a class="headerlink" href="#parameters_55" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix with CpG sites as rows and samples as columns.
    Values are typically beta (01) or M-values (-inf to +inf).</li>
<li><strong>pheno</strong>: <code>pd.DataFrame</code>
    Sample metadata table. Must be indexed by sample IDs (strings         after construction).</li>
<li><strong>ann</strong>: <code>pd.DataFrame or None</code>
    Optional probe/CpG annotation table (e.g., Illumina manifest).         Must be indexed by CpG IDs.
    If <code>None</code>, probe-level alignment checks are skipped.</li>
<li><strong>meta</strong>: <code>dict, optional</code>
    Free-form dictionary storing pipeline provenance and parameters.
    Pre-populated with sensible defaults if not provided.</li>
</ul>
<h3 id="attributes_1">Attributes<a class="headerlink" href="#attributes_1" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix (string-indexed rows and columns).</li>
<li><strong>pheno</strong>: <code>pd.DataFrame</code>
    Sample metadata (string-indexed).</li>
<li><strong>ann</strong>: <code>pd.DataFrame or None</code>
    Probe annotation (string-indexed if present).</li>
<li><strong>meta</strong>: <code>dict</code>
    Processing metadata (e.g., <code>{"matrix_type": "beta",         "normalized": True, ...}</code>).</li>
</ul>
<h3 id="notes_13">Notes<a class="headerlink" href="#notes_13" title="Permanent link">&para;</a></h3>
<p>The <code>__post_init__</code> method automatically:</p>
<ul>
<li>Converts all relevant indices to strings.</li>
<li>Validates alignment between components.</li>
<li>Raises informative <code>KeyError</code> on mismatch.</li>
</ul>
<h3 id="readers">Readers<a class="headerlink" href="#readers" title="Permanent link">&para;</a></h3>
<p>Input utilities for loading methylation datasets in the DMeth pipeline.</p>
<p>This module provides a <strong>Python API</strong> for loading
preprocessed or raw methylation data into the standard :class:<code>ProcessedData</code> container used throughout the package.</p>
<h3 id="features_12">Features<a class="headerlink" href="#features_12" title="Permanent link">&para;</a></h3>
<ul>
<li>Robust multi-format reading (CSV, TSV, Excel, pickle, HDF5, Parquet, Feather)</li>
<li>Automatic coercion of methylation matrix to numeric values</li>
<li>Smart HDF5 key selection (prefers <code>/M</code> or <code>M</code>)</li>
<li>Safe sample alignment between methylation matrix and phenotype data</li>
<li>Two high-level entry points:<ul>
<li><code>load_processed_data()</code>  load a previously saved <code>ProcessedData</code> object</li>
<li><code>load_methylation_data()</code>  construct a new <code>ProcessedData</code> from raw files</li>
</ul>
</li>
</ul>
<hr />
<p><code>ProcessedData(M: DataFrame, pheno: DataFrame, ann: DataFrame, meta: Dict[str, Any] = &lt;factory&gt;) -&gt; None</code></p>
<p>Central container for aligned methylation data used throughout the DMeth pipeline.</p>
<p>Guarantees that:</p>
<ul>
<li>All components (methylation matrix, sample metadata, probe annotation)     share consistent string indices.</li>
<li>Sample and probe alignment is validated on construction.</li>
<li>A <code>meta</code> dictionary tracks processing history (normalisation,     batch correction, QC metrics, etc.).</li>
</ul>
<p>All downstream functions expect or return instances of this class.</p>
<h3 id="parameters_56">Parameters<a class="headerlink" href="#parameters_56" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix with CpG sites as rows and samples as columns.
    Values are typically beta (01) or M-values (-inf to +inf).</li>
<li><strong>pheno</strong>: <code>pd.DataFrame</code>
    Sample metadata table. Must be indexed by sample IDs (strings         after construction).</li>
<li><strong>ann</strong>: <code>pd.DataFrame or None</code>
    Optional probe/CpG annotation table (e.g., Illumina manifest).         Must be indexed by CpG IDs.
    If <code>None</code>, probe-level alignment checks are skipped.</li>
<li><strong>meta</strong>: <code>dict, optional</code>
    Free-form dictionary storing pipeline provenance and parameters.
    Pre-populated with sensible defaults if not provided.</li>
</ul>
<h3 id="attributes_2">Attributes<a class="headerlink" href="#attributes_2" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix (string-indexed rows and columns).</li>
<li><strong>pheno</strong>: <code>pd.DataFrame</code>
    Sample metadata (string-indexed).</li>
<li><strong>ann</strong>: <code>pd.DataFrame or None</code>
    Probe annotation (string-indexed if present).</li>
<li><strong>meta</strong>: <code>dict</code>
    Processing metadata (e.g., <code>{"matrix_type": "beta",         "normalized": True, ...}</code>).</li>
</ul>
<h3 id="notes_14">Notes<a class="headerlink" href="#notes_14" title="Permanent link">&para;</a></h3>
<p>The <code>__post_init__</code> method automatically:</p>
<ul>
<li>Converts all relevant indices to strings.</li>
<li>Validates alignment between components.</li>
<li>Raises informative <code>KeyError</code> on mismatch.</li>
</ul>
<hr />
<p><code>load_methylation_data(methylation_input: Union[str, pathlib.Path, DataFrame], pheno_input: Union[str, pathlib.Path, DataFrame, NoneType] = None, ann_input: Union[str, pathlib.Path, DataFrame, NoneType] = None, index_col_probe: Optional[int] = 0, index_col_sample: Optional[int] = 0) -&gt; ProcessedData</code></p>
<p>High-level constructor that builds a ProcessedData object from     raw matrix/metadata files.</p>
<ul>
<li>Accepts file paths, already-loaded DataFrames, or a mix of both.</li>
<li>Performs automatic type coercion, sample alignment, and validation.</li>
</ul>
<h3 id="parameters_57">Parameters<a class="headerlink" href="#parameters_57" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>methylation_input</strong>: <code>str | Path | pd.DataFrame</code>
    CpG  sample methylation matrix (beta or M-values).</li>
<li><strong>pheno_input</strong>: <code>str | Path | pd.DataFrame, optional</code>
    Sample phenotype/metadata table. If <code>None</code>, an empty placeholder indexed by
    the matrix columns is created.</li>
<li><strong>ann_input</strong>: <code>str | Path | pd.DataFrame, optional</code>
    CpG probe annotation (e.g., Illumina manifest). If <code>None</code>,         annotation is omitted.</li>
<li><strong>index_col_probe</strong>: <code>int, default 0</code>
    Column containing CpG identifiers when reading the methylation matrix.</li>
<li><strong>index_col_sample</strong>: <code>int, default 0</code>
    Column containing sample identifiers when reading the phenotype table.</li>
</ul>
<h3 id="returns_50">Returns<a class="headerlink" href="#returns_50" title="Permanent link">&para;</a></h3>
<p>ProcessedData
    Fully validated container ready for downstream analysis.</p>
<h3 id="notes_15">Notes<a class="headerlink" href="#notes_15" title="Permanent link">&para;</a></h3>
<ul>
<li>Non-numeric entries in the methylation matrix are forcibly coerced to NaN
  (with a warning indicating how many values were affected).</li>
<li>The phenotype table is reindexed to exactly match the sample order in <code>M</code>.</li>
<li>All indices are converted to strings and alignment is checked by     <code>ProcessedData.__post_init__</code>.</li>
</ul>
<hr />
<p><code>load_processed_data(path: Union[str, pathlib.Path], trusted: bool = False) -&gt; ProcessedData</code></p>
<p>Load a previously saved ProcessedData object from disk.</p>
<p>Supports the two canonical DMeth output formats:
- <code>.pkl</code> / <code>.pickle</code>  direct pandas pickle of the ProcessedData instance
- <code>.h5</code> / <code>.hdf5</code>  HDF5 store containing the keys <code>M</code>,     <code>pheno</code>, and optionally <code>ann</code></p>
<h3 id="parameters_58">Parameters<a class="headerlink" href="#parameters_58" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>path</strong>: <code>str or Path</code>
    Path to the saved ProcessedData file.</li>
<li><strong>trusted</strong>: <code>bool, default False</code>
    Only allow <code>.pkl</code> / <code>.pickle</code> files from trusted sources</li>
</ul>
<h3 id="returns_51">Returns<a class="headerlink" href="#returns_51" title="Permanent link">&para;</a></h3>
<p>ProcessedData
    Fully reconstructed and validated ProcessedData container.</p>
<h3 id="raises_19">Raises<a class="headerlink" href="#raises_19" title="Permanent link">&para;</a></h3>
<p>ValueError
    If pickle file is from an untrusted source
ValueError
    If the file extension is not one of the supported formats.</p>
<h3 id="writers">Writers<a class="headerlink" href="#writers" title="Permanent link">&para;</a></h3>
<p>Output utilities for saving processed methylation data and analysis results.</p>
<p>This module provides a unified interface (Python API + Typer CLI) for exporting:</p>
<ul>
<li>ProcessedData objects in multiple formats</li>
<li>Differential methylation results (filtered and clean)</li>
<li>IDAT-style HDF5 archives (compatible with downstream tools)</li>
<li>Self-contained HTML analysis reports with embedded plots</li>
</ul>
<p>All export functions include overwrite protection, automatic directory creation,
and informative logging.</p>
<h3 id="features_13">Features<a class="headerlink" href="#features_13" title="Permanent link">&para;</a></h3>
<ul>
<li>Smart format detection and suffix handling</li>
<li>Optional compression (blosc:zstd or gzip)</li>
<li>Safe export of differential results (core columns + group means)</li>
<li>IDAT-compatible HDF5 layout via h5py (fallback to pandas HDFStore)</li>
<li>One-call HTML report generation from summary dict + matplotlib figures</li>
</ul>
<hr />
<p><code>ProcessedData(M: DataFrame, pheno: DataFrame, ann: DataFrame, meta: Dict[str, Any] = &lt;factory&gt;) -&gt; None</code></p>
<p>Central container for aligned methylation data used throughout the DMeth pipeline.</p>
<p>Guarantees that:</p>
<ul>
<li>All components (methylation matrix, sample metadata, probe annotation)     share consistent string indices.</li>
<li>Sample and probe alignment is validated on construction.</li>
<li>A <code>meta</code> dictionary tracks processing history (normalisation,     batch correction, QC metrics, etc.).</li>
</ul>
<p>All downstream functions expect or return instances of this class.</p>
<h3 id="parameters_59">Parameters<a class="headerlink" href="#parameters_59" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix with CpG sites as rows and samples as columns.
    Values are typically beta (01) or M-values (-inf to +inf).</li>
<li><strong>pheno</strong>: <code>pd.DataFrame</code>
    Sample metadata table. Must be indexed by sample IDs (strings         after construction).</li>
<li><strong>ann</strong>: <code>pd.DataFrame or None</code>
    Optional probe/CpG annotation table (e.g., Illumina manifest).         Must be indexed by CpG IDs.
    If <code>None</code>, probe-level alignment checks are skipped.</li>
<li><strong>meta</strong>: <code>dict, optional</code>
    Free-form dictionary storing pipeline provenance and parameters.
    Pre-populated with sensible defaults if not provided.</li>
</ul>
<h3 id="attributes_3">Attributes<a class="headerlink" href="#attributes_3" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix (string-indexed rows and columns).</li>
<li><strong>pheno</strong>: <code>pd.DataFrame</code>
    Sample metadata (string-indexed).</li>
<li><strong>ann</strong>: <code>pd.DataFrame or None</code>
    Probe annotation (string-indexed if present).</li>
<li><strong>meta</strong>: <code>dict</code>
    Processing metadata (e.g., <code>{"matrix_type": "beta",         "normalized": True, ...}</code>).</li>
</ul>
<h3 id="notes_16">Notes<a class="headerlink" href="#notes_16" title="Permanent link">&para;</a></h3>
<p>The <code>__post_init__</code> method automatically:</p>
<ul>
<li>Converts all relevant indices to strings.</li>
<li>Validates alignment between components.</li>
<li>Raises informative <code>KeyError</code> on mismatch.</li>
</ul>
<hr />
<p><code>create_analysis_report(summary: Dict[str, Any], plots: Dict[str, matplotlib.figure.Figure], outpath: Union[str, pathlib.Path] = 'analysis_report.html', title: str = 'Methylation Analysis Report') -&gt; pathlib.Path</code></p>
<p>Generate a self-contained, minimal HTML report embedding a JSON     summary and PNG images of supplied figures.</p>
<p>Useful for quick sharing of results without requiring Jupyter or     additional dependencies.</p>
<h3 id="parameters_60">Parameters<a class="headerlink" href="#parameters_60" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>summary</strong>: <code>dict</code>
    Arbitrary summary statistics or results (JSON-serialisable).</li>
<li><strong>plots</strong>: <code>dict[str, plt.Figure]</code>
    Mapping from plot name to Matplotlib figure objects.</li>
<li><strong>outpath</strong>: <code>str or Path, default "analysis_report.html"</code>
    Output HTML file location.</li>
<li><strong>title</strong>: <code>str, default "Methylation Analysis Report"</code>
    Top-level heading in the report.</li>
</ul>
<h3 id="returns_52">Returns<a class="headerlink" href="#returns_52" title="Permanent link">&para;</a></h3>
<p>Path
    Path to the generated HTML file.</p>
<hr />
<p><code>export_idat_hdf5(beta: DataFrame, mvals: Optional[DataFrame], sample_sheet: Optional[DataFrame], filepath: Union[str, pathlib.Path], compress: bool = True) -&gt; pathlib.Path</code></p>
<p>Write methylation data to an IDAT-compatible HDF5 archive using h5py     (preferred) or pandas fallback.</p>
<p>Layout mirrors common downstream tools: groups <code>/beta</code>, <code>/mvals</code> (optional),
and <code>/samples</code> containing one JSON dataset per sample.</p>
<h3 id="parameters_61">Parameters<a class="headerlink" href="#parameters_61" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>beta</strong>: <code>pd.DataFrame</code>
    Beta-value matrix (CpGs  samples).</li>
<li><strong>mvals</strong>: <code>pd.DataFrame or None, optional</code>
    M-value matrix (same dimensions).</li>
<li><strong>sample_sheet</strong>: <code>pd.DataFrame or None, optional</code>
    Sample metadata (index = sample IDs).</li>
<li><strong>filepath</strong>: <code>str or Path</code>
    Destination HDF5 file.</li>
<li><strong>compress</strong>: <code>bool, default True</code>
    Apply gzip compression to large datasets.</li>
</ul>
<h3 id="returns_53">Returns<a class="headerlink" href="#returns_53" title="Permanent link">&para;</a></h3>
<p>Path
    Path to the created HDF5 file.</p>
<h3 id="raises_20">Raises<a class="headerlink" href="#raises_20" title="Permanent link">&para;</a></h3>
<p>RuntimeError
    If neither h5py nor pandas.HDFStore is available.</p>
<hr />
<p><code>export_results(res: DataFrame, output_path: str, format: str = 'csv', include_all: bool = False, verbose: bool = True, overwrite: bool = True) -&gt; None</code></p>
<p>Export differential methylation results to CSV, TSV, or Excel with     optional column filtering.</p>
<ul>
<li>By default only core statistical columns and per-group mean columns are exported.</li>
<li>Overwrite protection is applied if <code>overwrite=False</code>.</li>
</ul>
<h3 id="parameters_62">Parameters<a class="headerlink" href="#parameters_62" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>res</strong>: <code>pd.DataFrame</code>
    Differential results table (as returned by the analysis pipeline).</li>
<li><strong>output_path</strong>: <code>str</code>
    Destination file path (extension determines format if <code>format</code> is omitted).</li>
<li><strong>format</strong>: <code>{"csv", "tsv", "excel"}, default "csv"</code>
    Output file format.</li>
<li><strong>include_all</strong>: <code>bool, default False</code>
    If True, export every column; if False, only core + mean columns are kept.</li>
<li><strong>verbose</strong>: <code>bool, default True</code>
    Log a confirmation message on success.</li>
<li><strong>overwrite</strong>: <code>bool, default True</code>
    Raise FileExistsError if the file exists and this is False.</li>
</ul>
<h3 id="raises_21">Raises<a class="headerlink" href="#raises_21" title="Permanent link">&para;</a></h3>
<p>FileExistsError
    When <code>overwrite=False</code> and the target file already exists.
ValueError
    For unsupported <code>format</code>.
RuntimeError
    If writing fails (e.g., missing openpyxl for Excel).</p>
<hr />
<p><code>save_processed_data(data: 'ProcessedData', path: Union[str, pathlib.Path], format: str = 'hdf5', compression: str | None = 'blosc:zstd', complib_level: int = 9, overwrite: bool = True) -&gt; None</code></p>
<p>Persist a ProcessedData object to disk in one of several convenient formats.</p>
<h2 id="supported-formats">Supported formats<a class="headerlink" href="#supported-formats" title="Permanent link">&para;</a></h2>
<p>csv / tsv       three separate files (M, pheno, ann)
xlsx            single workbook with sheets "M", "pheno", "ann"
pickle / pkl    direct pickle of the ProcessedData instance
hdf5 / h5       pandas HDFStore with compressed tables</p>
<h3 id="parameters_63">Parameters<a class="headerlink" href="#parameters_63" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>data</strong>: <code>ProcessedData</code>
    The container to save.</li>
<li><strong>path</strong>: <code>str or Path</code>
    Output path; suffix is automatically corrected to match <code>format</code>.</li>
<li><strong>format</strong>: <code>{"csv", "tsv", "xlsx", "pickle", "pkl", "hdf5", "h5"}, default "hdf5"</code>
    Desired export format.</li>
<li><strong>compression</strong>: <code>str or None, default "blosc:zstd"</code>
    Compression library for HDF5 (only used when format is hdf5).</li>
<li><strong>complib_level</strong>: <code>int, default 9</code>
    Compression level (19) for HDF5.</li>
<li><strong>overwrite</strong>: <code>bool, default True</code>
    Allow overwriting existing files.</li>
</ul>
<h3 id="raises_22">Raises<a class="headerlink" href="#raises_22" title="Permanent link">&para;</a></h3>
<p>ValueError
    If an unsupported format is requested.
FileExistsError
    When <code>overwrite=False</code> and the target exists.</p>
<hr />
<h2 id="utilities">Utilities<a class="headerlink" href="#utilities" title="Permanent link">&para;</a></h2>
<h3 id="logger">Logger<a class="headerlink" href="#logger" title="Permanent link">&para;</a></h3>
<p>Centralised logging utilities for the DMeth DNA-methylation analysis suite.</p>
<p>This module configures a unified logger with the following features:</p>
<h3 id="features_14">Features<a class="headerlink" href="#features_14" title="Permanent link">&para;</a></h3>
<ul>
<li>Timestamped log files automatically saved to <code>&lt;output_dir&gt;/log/</code></li>
<li>Simultaneous console (stdout) and file output</li>
<li>Consistent formatting across all package modules</li>
<li>A custom :class:<code>ProgressAwareLogger</code> that seamlessly integrates <code>tqdm</code> progress bars:<ul>
<li><code>logger.progress("Processing samples", total=n)</code> starts a progress bar</li>
<li><code>logger.progress_update(k)</code> advances it</li>
<li>Any regular log call (info/warning/error/etc.) automatically closes the active     bar so that log messages are never corrupted by overlapping tqdm output</li>
</ul>
</li>
</ul>
<p>All other <code>dmeth</code> modules import the logger via <code>get_logger()</code>.</p>
<hr />
<p><code>ProgressAwareLogger(name)</code></p>
<p>Custom logger class that supports a temporary progress bar.
The progress bar stays active until the next normal log call.</p>
<hr />
<p><code>get_logger(name: str = 'dmeth') -&gt; logging.Logger</code></p>
<p>Return the central DMeth logger instance.</p>
<h3 id="parameters_64">Parameters<a class="headerlink" href="#parameters_64" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>name</strong>: <code>str, default "dmeth"</code>
    Logger name. Typically left as the default.</li>
</ul>
<h3 id="returns_54">Returns<a class="headerlink" href="#returns_54" title="Permanent link">&para;</a></h3>
<p>logging.Logger
    The configured ProgressAwareLogger instance.</p>
<h3 id="plotting">Plotting<a class="headerlink" href="#plotting" title="Permanent link">&para;</a></h3>
<p>Plotting utilities for downstream DNA methylation analysis.</p>
<p>This module provides a comprehensive set of visualisation functions for quality control,
differential methylation analysis, dimensionality reduction, variance shrinkage diagnostics,
and publication-ready summary figures. Functions operate on pandas DataFrames, ProcessedData
objects, or differential methylation results and return standardised matplotlib (or optionally
Plotly) figures.</p>
<h3 id="features_15">Features<a class="headerlink" href="#features_15" title="Permanent link">&para;</a></h3>
<ul>
<li>Standardised figure creation</li>
<li>Enhanced volcano plots with automatic significance colouring and annotation</li>
<li>P-value QQ plots and histograms</li>
<li>Sample-level PCA, t-SNE, and UMAP embeddings (static or interactive)</li>
<li>Variance shrinkage diagnostics (limma-style)</li>
<li>Mean-difference bar plots for top loci</li>
<li>Multi-panel QC reports via plot_stage()</li>
<li>Combined DMS/DMR summary visualisation (volcano + manhattan + heatmap)</li>
</ul>
<hr />
<p><code>ProcessedData(M: DataFrame, pheno: DataFrame, ann: DataFrame, meta: Dict[str, Any] = &lt;factory&gt;) -&gt; None</code></p>
<p>Central container for aligned methylation data used throughout the DMeth pipeline.</p>
<p>Guarantees that:</p>
<ul>
<li>All components (methylation matrix, sample metadata, probe annotation)     share consistent string indices.</li>
<li>Sample and probe alignment is validated on construction.</li>
<li>A <code>meta</code> dictionary tracks processing history (normalisation,     batch correction, QC metrics, etc.).</li>
</ul>
<p>All downstream functions expect or return instances of this class.</p>
<h3 id="parameters_65">Parameters<a class="headerlink" href="#parameters_65" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix with CpG sites as rows and samples as columns.
    Values are typically beta (01) or M-values (-inf to +inf).</li>
<li><strong>pheno</strong>: <code>pd.DataFrame</code>
    Sample metadata table. Must be indexed by sample IDs (strings         after construction).</li>
<li><strong>ann</strong>: <code>pd.DataFrame or None</code>
    Optional probe/CpG annotation table (e.g., Illumina manifest).         Must be indexed by CpG IDs.
    If <code>None</code>, probe-level alignment checks are skipped.</li>
<li><strong>meta</strong>: <code>dict, optional</code>
    Free-form dictionary storing pipeline provenance and parameters.
    Pre-populated with sensible defaults if not provided.</li>
</ul>
<h3 id="attributes_4">Attributes<a class="headerlink" href="#attributes_4" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix (string-indexed rows and columns).</li>
<li><strong>pheno</strong>: <code>pd.DataFrame</code>
    Sample metadata (string-indexed).</li>
<li><strong>ann</strong>: <code>pd.DataFrame or None</code>
    Probe annotation (string-indexed if present).</li>
<li><strong>meta</strong>: <code>dict</code>
    Processing metadata (e.g., <code>{"matrix_type": "beta",         "normalized": True, ...}</code>).</li>
</ul>
<h3 id="notes_17">Notes<a class="headerlink" href="#notes_17" title="Permanent link">&para;</a></h3>
<p>The <code>__post_init__</code> method automatically:</p>
<ul>
<li>Converts all relevant indices to strings.</li>
<li>Validates alignment between components.</li>
<li>Raises informative <code>KeyError</code> on mismatch.</li>
</ul>
<hr />
<p><code>correlate_methylation_expression(beta: 'pd.DataFrame', expression: 'pd.DataFrame', gene_map: 'Optional[Dict[str, str]]' = None, method: 'str' = 'pearson', override_index_alignment: 'bool' = False) -&gt; 'pd.DataFrame'</code></p>
<p>Compute sample-wise correlation between CpG methylation and gene expression.</p>
<p>Supports both one-to-one (same index) and many-to-one (custom CpGgene     mapping) scenarios.</p>
<h3 id="parameters_66">Parameters<a class="headerlink" href="#parameters_66" title="Permanent link">&para;</a></h3>
<p>beta, expression : pd.DataFrame
    Methylation (CpGs  samples) and expression (genes  samples) matrices.
- <strong>gene_map</strong>: <code>dict or None</code>
    Explicit mapping from CpG ID  gene symbol (required for cis-analysis).
- <strong>method</strong>: <code>{"pearson", "spearman"}, default "pearson"</code>
    Correlation coefficient to compute.
- <strong>override_index_alignment</strong>: <code>bool, default False</code>
    Proceed even with &lt;2 overlapping samples (useful for exploratory checks).</p>
<h3 id="returns_55">Returns<a class="headerlink" href="#returns_55" title="Permanent link">&para;</a></h3>
<p>pd.DataFrame
    Correlation results with columns <code>r</code> and <code>pval</code> (indexed by CpG         and optionally gene).</p>
<hr />
<p><code>methylation_expression_heatmap(beta: DataFrame, expression: DataFrame, genes: Optional[List[str]] = None, top_n: int = 50, sample_metadata: Optional[DataFrame] = None, save_path: Union[str, pathlib.Path, NoneType] = None, method: str = 'pearson') -&gt; matplotlib.figure.Figure</code></p>
<p>Heatmap of methylation-expression correlation coefficients for selected genes/CpGs.</p>
<h3 id="parameters_67">Parameters<a class="headerlink" href="#parameters_67" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>beta</strong>: <code>pd.DataFrame</code>
    Methylation beta matrix (features  samples).</li>
<li><strong>expression</strong>: <code>pd.DataFrame</code>
    Gene expression matrix (genes  samples).</li>
<li><strong>genes</strong>: <code>list[str], optional</code>
    Specific features to display; if None the top_n most correlated are used.</li>
<li><strong>top_n</strong>: <code>int, default 50</code>
    Number of top correlations shown when <code>genes</code> is None.</li>
<li><strong>sample_metadata</strong>: <code>pd.DataFrame, optional</code>
    Currently unused (reserved for future sample-side annotation).</li>
<li><strong>save_path</strong>: <code>str or Path, optional</code>
    Destination for saving the figure.</li>
<li><strong>method</strong>: <code>{"pearson", "spearman"}, default "pearson"</code>
    Correlation method.</li>
</ul>
<h3 id="returns_56">Returns<a class="headerlink" href="#returns_56" title="Permanent link">&para;</a></h3>
<p>plt.Figure
    Horizontal heatmap of correlation coefficients.</p>
<hr />
<p><code>pca_plot(data: ProcessedData, color_col: str, title: str = 'PCA of Methylation Data', figsize: tuple[int, int] = (7, 6)) -&gt; None</code></p>
<p>Display a simple PCA projection of samples coloured by a phenotype column.</p>
<h3 id="parameters_68">Parameters<a class="headerlink" href="#parameters_68" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>data</strong>: <code>ProcessedData</code>
    Container holding the methylation matrix and phenotype table.</li>
<li><strong>color_col</strong>: <code>str</code>
    Column in <code>data.pheno</code> used for colouring points.</li>
<li><strong>title</strong>: <code>str, default "PCA of Methylation Data"</code>
    Title shown on the plot.</li>
<li><strong>figsize</strong>: <code>tuple[int, int], default (7, 6)</code>
    Size of the figure in inches.</li>
</ul>
<h3 id="notes_18">Notes<a class="headerlink" href="#notes_18" title="Permanent link">&para;</a></h3>
<p>The plot is shown immediately with <code>plt.show()</code> and is not returned.</p>
<hr />
<p><code>plot_mean_difference(beta_group1: DataFrame, beta_group2: DataFrame, top_n: int = 50, save_path: Union[str, pathlib.Path, NoneType] = None) -&gt; matplotlib.figure.Figure</code></p>
<p>Bar plot of the largest absolute mean beta differences between two groups.</p>
<h3 id="parameters_69">Parameters<a class="headerlink" href="#parameters_69" title="Permanent link">&para;</a></h3>
<p>beta_group1, beta_group2 : pd.DataFrame
    Beta-value matrices (features  samples) for each group.
- <strong>top_n</strong>: <code>int, default 50</code>
    Number of top differentially methylated loci to display.
- <strong>save_path</strong>: <code>str or Path, optional</code>
    Destination for saving the figure.</p>
<h3 id="returns_57">Returns<a class="headerlink" href="#returns_57" title="Permanent link">&para;</a></h3>
<p>plt.Figure
    Bar chart of absolute mean differences.</p>
<hr />
<p><code>plot_pvalue_qq(res: DataFrame, pval_col: str = 'pval', dpi: int = 300, save_path: Union[str, pathlib.Path, NoneType] = None) -&gt; matplotlib.figure.Figure</code></p>
<p>Produce a quantile-quantile plot comparing observed versus     expected -log10(p-values).</p>
<h3 id="parameters_70">Parameters<a class="headerlink" href="#parameters_70" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>res</strong>: <code>pd.DataFrame</code>
    DataFrame containing the p-value column.</li>
<li><strong>pval_col</strong>: <code>str, default "pval"</code>
    Column name of the p-values.</li>
<li><strong>dpi</strong>: <code>int, default 300</code>
    Figure resolution.</li>
<li><strong>save_path</strong>: <code>str or Path, optional</code>
    Destination path for saving the figure.</li>
</ul>
<h3 id="returns_58">Returns<a class="headerlink" href="#returns_58" title="Permanent link">&para;</a></h3>
<p>plt.Figure
    Q-Q plot figure.</p>
<hr />
<p><code>plot_shrinkage_diagnostics(s2: pandas.core.series.Series, s2_post: pandas.core.series.Series, d0: Optional[float] = None, save_path: Union[str, pathlib.Path, NoneType] = None) -&gt; matplotlib.figure.Figure</code></p>
<p>Visualise the effect of empirical Bayes variance shrinkage.</p>
<h3 id="parameters_71">Parameters<a class="headerlink" href="#parameters_71" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>s2</strong>: <code>pd.Series</code>
    Original (unmoderated) variance estimates per feature.</li>
<li><strong>s2_post</strong>: <code>pd.Series</code>
    Moderated (shrunk) variance estimates.</li>
<li><strong>d0</strong>: <code>float, optional</code>
    Prior degrees of freedom from the shrinkage procedure (displayed         in title if given).</li>
<li><strong>save_path</strong>: <code>str or Path, optional</code>
    If provided, the figure is saved to this location.</li>
</ul>
<h3 id="returns_59">Returns<a class="headerlink" href="#returns_59" title="Permanent link">&para;</a></h3>
<p>plt.Figure
    Scatter plot comparing log10(original) vs log10(shrunk) variances.</p>
<hr />
<p><code>plot_stage(stage: str, M: DataFrame, res: Optional[DataFrame] = None, metadata: Optional[DataFrame] = None, groups_col: str = 'Type', top_n: int = 10, embedding: str = 'pca', interactive: bool = False, save_path: Union[str, Dict[str, str], NoneType] = None, dpi: int = 300, **kwargs) -&gt; Dict[str, Union[matplotlib.figure.Figure, ForwardRef('plotly.graph_objects.Figure')]]</code></p>
<p>Orchestrate a complete multi-panel QC or analysis visualisation for     a given processing stage.</p>
<p>Supported stages: "qc", "variance", "differential", "top_hits", "correlation".</p>
<h3 id="parameters_72">Parameters<a class="headerlink" href="#parameters_72" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>stage</strong>: <code>str</code>
    One of the recognised analysis stages.</li>
<li><strong>M</strong>: <code>pd.DataFrame</code>
    Methylation matrix (features  samples), typically M-values or beta-values.</li>
<li><strong>res</strong>: <code>pd.DataFrame, optional</code>
    Differential methylation results (required for variance,         differential, top_hits stages).</li>
<li><strong>metadata</strong>: <code>pd.DataFrame, optional</code>
    Sample annotation table; must contain the grouping column.</li>
<li><strong>groups_col</strong>: <code>str, default "Type"</code>
    Column in metadata used for colour-coding samples.</li>
<li><strong>top_n</strong>: <code>int, default 10</code>
    Number of top features to display in relevant panels.</li>
<li><strong>embedding</strong>: <code>{"pca", "tsne", "umap"}, default "pca"</code>
    Dimensionality-reduction method for sample embedding.</li>
<li><strong>interactive</strong>: <code>bool, default False</code>
    Return Plotly interactive figures where possible.</li>
<li><strong>save_path</strong>: <code>str, Path or dict, optional</code>
    Path or mapping of figure names to paths for automatic saving.</li>
<li><strong>dpi</strong>: <code>int, default 300</code>
    Resolution for static figures.
**kwargs
    Additional keyword arguments passed to the underlying reducer.</li>
</ul>
<h3 id="returns_60">Returns<a class="headerlink" href="#returns_60" title="Permanent link">&para;</a></h3>
<p>dict[str, plt.Figure | plotly.graph_objects.Figure]
    Mapping from panel name to the generated figure(s).</p>
<hr />
<p><code>plot_volcano(res: DataFrame, lfc_col: str = 'logFC', pval_col: str = 'pval', alpha: float = 0.7, lfc_thresh: float = 1.0, pval_thresh: float = 0.05, top_n: int = 10, dpi: int = 300, save_path: Union[str, pathlib.Path, NoneType] = None) -&gt; matplotlib.figure.Figure</code></p>
<p>Create an enhanced volcano plot with top hits annotated.</p>
<h3 id="parameters_73">Parameters<a class="headerlink" href="#parameters_73" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>res</strong>: <code>pd.DataFrame</code>
    Differential results with logFC and p-values</li>
<li><strong>lfc_col</strong>: <code>str</code>
    Column name for log fold change</li>
<li><strong>pval_col</strong>: <code>str</code>
    Column name for p-value</li>
<li><strong>alpha</strong>: <code>float</code>
    Point transparency</li>
<li><strong>lfc_thresh</strong>: <code>float</code>
    Threshold for logFC significance</li>
<li><strong>pval_thresh</strong>: <code>float</code>
    Threshold for p-value significance</li>
<li><strong>top_n</strong>: <code>int</code>
    Number of top hits to annotate</li>
<li><strong>dpi</strong>: <code>int</code>
    Figure resolution</li>
<li><strong>save_path</strong>: <code>str or Path, optional</code>
    Path to save figure</li>
</ul>
<h3 id="returns_61">Returns<a class="headerlink" href="#returns_61" title="Permanent link">&para;</a></h3>
<p>plt.Figure
    Volcano plot figure</p>
<hr />
<p><code>pvalue_histogram(pvals: Union[pandas.core.series.Series, ndarray, List[float]], bins: int = 50, save_path: Union[str, pathlib.Path, NoneType] = None) -&gt; matplotlib.figure.Figure</code></p>
<p>Histogram of p-values with a flat null-expectation line for QC assessment.</p>
<h3 id="parameters_74">Parameters<a class="headerlink" href="#parameters_74" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>pvals</strong>: <code>array-like</code>
    Collection of p-values.</li>
<li><strong>bins</strong>: <code>int, default 50</code>
    Number of histogram bins.</li>
<li><strong>save_path</strong>: <code>str or Path, optional</code>
    Path to save the figure.</li>
</ul>
<h3 id="returns_62">Returns<a class="headerlink" href="#returns_62" title="Permanent link">&para;</a></h3>
<p>plt.Figure
    Histogram figure.</p>
<hr />
<p><code>visualize_dms(res: DataFrame, beta: Optional[DataFrame] = None, top_n: int = 50, volcano: bool = True, manhattan: bool = True, heatmap: bool = True, sample_metadata: Optional[DataFrame] = None, save_dir: Union[str, pathlib.Path, NoneType] = None) -&gt; Dict[str, Optional[matplotlib.figure.Figure]]</code></p>
<p>Produce a standard set of summary plots for differentially methylated sites/regions.</p>
<p>Optionally creates volcano, Manhattan, and heatmap visualisations.</p>
<h3 id="parameters_75">Parameters<a class="headerlink" href="#parameters_75" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>res</strong>: <code>pd.DataFrame</code>
    Differential methylation results.</li>
<li><strong>beta</strong>: <code>pd.DataFrame, optional</code>
    Beta matrix required for the heatmap panel.</li>
<li><strong>top_n</strong>: <code>int, default 50</code>
    Number of top CpGs shown in the heatmap.
volcano / manhattan / heatmap : bool, default True
    Toggle creation of each panel.</li>
<li><strong>sample_metadata</strong>: <code>pd.DataFrame, optional</code>
    Sample annotation (currently unused but reserved for future clustering).</li>
<li><strong>save_dir</strong>: <code>str or Path, optional</code>
    Directory where individual PNG files are written.</li>
</ul>
<h3 id="returns_63">Returns<a class="headerlink" href="#returns_63" title="Permanent link">&para;</a></h3>
<p>dict[str, plt.Figure | None]
    Figures for "volcano", "manhattan", and "heatmap" (None if not generated).</p>
<h2 id="citation">Citation<a class="headerlink" href="#citation" title="Permanent link">&para;</a></h2>
<p>If you use <code>dmeth</code> in your research, please cite:</p>
<div class="codehilite"><pre><span></span><code><span class="nc">@software</span><span class="p">{</span><span class="nl">dmeth2025</span><span class="p">,</span>
<span class="w">  </span><span class="na">author</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Afolabi, Dare}</span><span class="p">,</span>
<span class="w">  </span><span class="na">title</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{dmeth: A comprehensive Python toolkit for differential DNA methylation analysis with empirical Bayes moderation and biomarker discovery}</span><span class="p">,</span>
<span class="w">  </span><span class="na">version</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{0.1.1}</span><span class="p">,</span>
<span class="w">  </span><span class="na">year</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{2025}</span><span class="p">,</span>
<span class="w">  </span><span class="na">publisher</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{GitHub}</span><span class="p">,</span>
<span class="w">  </span><span class="na">url</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{https://github.com/dare-afolabi/dmeth}</span><span class="p">,</span>
<span class="w">  </span><span class="na">doi</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{10.5281/zenodo.17684038}</span><span class="p">,</span>
<span class="p">}</span>
</code></pre></div>

<h3 id="references">References<a class="headerlink" href="#references" title="Permanent link">&para;</a></h3>
<ul>
<li>Smyth, G. K. (2004). Linear models and empirical bayes methods for assessing differential expression in microarray experiments. <em>Statistical Applications in Genetics and Molecular Biology</em>, 3(1).</li>
<li>Liu, P., &amp; Hwang, J.T.G. (2007). Quick calculation for sample size while controlling false discovery rate with application to microarray analysis. <em>Bioinformatics</em>, 23(6), 739746.</li>
<li>Du, P., Zhang, X., Huang, C.-C., Jafari, N., Kibbe, W.A., Hou, L., &amp; Lin, S. (2010). Comparison of Beta-value and M-value methods for quantifying methylation levels by microarray analysis. <em>BMC Bioinformatics</em>, 11:587.</li>
<li>Jung, S.H., Young, S.S. (2012). Power and sample size calculation for microarray studies. <em>Journal of Biopharmaceutical Statistics</em>, 22(1):30-42.</li>
<li>Phipson, B. et al. (2016). missMethyl: an R package for analyzing data from Illuminas HumanMethylation450 platform. <em>Bioinformatics</em>, 32(2), 286-288.</li>
</ul>
<h2 id="support">Support<a class="headerlink" href="#support" title="Permanent link">&para;</a></h2>
<ul>
<li><strong>Issues</strong>: <a href="https://github.com/dare-afolabi/dmeth/issues">GitHub Issues</a></li>
<li><strong>Discussions</strong>: <a href="https://github.com/dare-afolabi/dmeth/discussions/1">GitHub Discussions</a></li>
<li><strong>Email</strong>: <a href="mailto:dare.afolabi@outlook.com">dare.afolabi@outlook.com</a></li>
</ul>
<hr />
<blockquote>
<p><strong>Auto-generated</strong> on 2025-11-23 03:47:48</p>
</blockquote>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["navigation.sections", "content.code.copy", "content.code.annotate"], "search": "../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
    
  </body>
</html>